// @angular-material-extensions/pages/fesm5/angular-material-extensions-pages.js
import { __decorate, __metadata } from 'tslib';
import { ViewChild, TemplateRef, Input, Component, ViewEncapsulation, ChangeDetectionStrategy, ContentChild, EventEmitter, Renderer2, ElementRef, ContentChildren, QueryList, Output, HostListener, NgModule } from '@angular/core';
import { MatSidenav, MatSidenavModule } from '@angular/material/sidenav';
import { MatTabGroup, MatTabsModule } from '@angular/material/tabs';
import { CommonModule } from '@angular/common';
import { MatButtonModule } from '@angular/material/button';
import { MatCardModule } from '@angular/material/card';
import { MatIconModule } from '@angular/material/icon';
import { MatStepperModule } from '@angular/material/stepper';
import { MatToolbarModule } from '@angular/material/toolbar';

import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/material/card';
import * as ɵngcc2 from '@angular/material/sidenav';
import * as ɵngcc3 from '@angular/material/toolbar';
import * as ɵngcc4 from '@angular/material/stepper';
import * as ɵngcc5 from '@angular/common';
import * as ɵngcc6 from '@angular/material/button';
import * as ɵngcc7 from '@angular/material/icon';
import * as ɵngcc8 from '@angular/material/tabs';

function MatPageContentComponent_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "mat-card", 0);
    ɵngcc0.ɵɵelementStart(1, "mat-card-title");
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "mat-card-content");
    ɵngcc0.ɵɵprojection(4);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ctx_r0.title);
} }
var _c0 = ["*"];
function MatPageOutlookComponent_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵprojection(0);
} }
var _c1 = ["sidenav"];
function MatPagesComponent_mat_step_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "mat-step", 15);
    ɵngcc0.ɵɵelementContainer(1, 16);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var page_r8 = ctx.$implicit;
    ɵngcc0.ɵɵproperty("label", page_r8.pageOutlook.title);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", page_r8.pageOutlook.content);
} }
function MatPagesComponent_mat_tab_17_ng_template_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0, 16);
} if (rf & 2) {
    var page_r10 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", page_r10.pageContent.content);
} }
function MatPagesComponent_mat_tab_17_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "mat-tab");
    ɵngcc0.ɵɵtemplate(1, MatPagesComponent_mat_tab_17_ng_template_1_Template, 1, 1, "ng-template", 17);
    ɵngcc0.ɵɵelementEnd();
} }
function MatPagesComponent_div_22_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "div");
} }
function MatPagesComponent_ng_template_23_Template(rf, ctx) { if (rf & 1) {
    var _r15 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 18);
    ɵngcc0.ɵɵlistener("click", function MatPagesComponent_ng_template_23_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r15); var ctx_r14 = ɵngcc0.ɵɵnextContext(); return ctx_r14.next(); });
    ɵngcc0.ɵɵelementStart(1, "mat-icon");
    ɵngcc0.ɵɵtext(2, "chevron_right");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
function MatPagesComponent_ng_template_25_Template(rf, ctx) { if (rf & 1) {
    var _r17 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 19);
    ɵngcc0.ɵɵlistener("click", function MatPagesComponent_ng_template_25_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r17); var ctx_r16 = ɵngcc0.ɵɵnextContext(); return ctx_r16.done.next(); });
    ɵngcc0.ɵɵelementStart(1, "mat-icon");
    ɵngcc0.ɵɵtext(2, "done");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
var MatPageContentComponent = /** @class */ (function () {
    function MatPageContentComponent() {
    }
    MatPageContentComponent.prototype.ngOnInit = function () {
    };
    __decorate([
        ViewChild(TemplateRef),
        __metadata("design:type", TemplateRef)
    ], MatPageContentComponent.prototype, "content", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], MatPageContentComponent.prototype, "title", void 0);
    MatPageContentComponent = __decorate([ __metadata("design:paramtypes", [])
    ], MatPageContentComponent);
MatPageContentComponent.ɵfac = function MatPageContentComponent_Factory(t) { return new (t || MatPageContentComponent)(); };
MatPageContentComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: MatPageContentComponent, selectors: [["mat-page-content"]], viewQuery: function MatPageContentComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(TemplateRef, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.content = _t.first);
    } }, inputs: { title: "title" }, ngContentSelectors: _c0, decls: 1, vars: 0, consts: [[2, "margin", "2rem"]], template: function MatPageContentComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵtemplate(0, MatPageContentComponent_ng_template_0_Template, 5, 1, "ng-template");
    } }, directives: [ɵngcc1.MatCard, ɵngcc1.MatCardTitle, ɵngcc1.MatCardContent], styles: [""] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(MatPageContentComponent, [{
        type: Component,
        args: [{
                selector: 'mat-page-content',
                template: "<ng-template>\n  <mat-card style=\"margin: 2rem\">\n    <mat-card-title>{{title}}</mat-card-title>\n    <mat-card-content>\n      <ng-content></ng-content>\n    </mat-card-content>\n  </mat-card>\n</ng-template>\n",
                styles: [""]
            }]
    }], function () { return []; }, { content: [{
            type: ViewChild,
            args: [TemplateRef]
        }], title: [{
            type: Input
        }] }); })();
    return MatPageContentComponent;
}());

var MatPageOutlookComponent = /** @class */ (function () {
    function MatPageOutlookComponent() {
    }
    MatPageOutlookComponent.prototype.ngOnInit = function () {
    };
    __decorate([
        ViewChild(TemplateRef),
        __metadata("design:type", TemplateRef)
    ], MatPageOutlookComponent.prototype, "content", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], MatPageOutlookComponent.prototype, "title", void 0);
    MatPageOutlookComponent = __decorate([ __metadata("design:paramtypes", [])
    ], MatPageOutlookComponent);
MatPageOutlookComponent.ɵfac = function MatPageOutlookComponent_Factory(t) { return new (t || MatPageOutlookComponent)(); };
MatPageOutlookComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: MatPageOutlookComponent, selectors: [["mat-page-outlook"]], viewQuery: function MatPageOutlookComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(TemplateRef, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.content = _t.first);
    } }, inputs: { title: "title" }, ngContentSelectors: _c0, decls: 1, vars: 0, template: function MatPageOutlookComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵtemplate(0, MatPageOutlookComponent_ng_template_0_Template, 1, 0, "ng-template");
    } }, styles: [""], encapsulation: 2, changeDetection: 0 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(MatPageOutlookComponent, [{
        type: Component,
        args: [{
                selector: 'mat-page-outlook',
                template: "<ng-template>\n  <ng-content></ng-content>\n</ng-template>\n",
                encapsulation: ViewEncapsulation.None,
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [""]
            }]
    }], function () { return []; }, { content: [{
            type: ViewChild,
            args: [TemplateRef]
        }], title: [{
            type: Input
        }] }); })();
    return MatPageOutlookComponent;
}());

var MatPageLoaderComponent = /** @class */ (function () {
    function MatPageLoaderComponent() {
    }
    __decorate([
        ContentChild(MatPageOutlookComponent),
        __metadata("design:type", MatPageOutlookComponent)
    ], MatPageLoaderComponent.prototype, "pageOutlook", void 0);
    __decorate([
        ContentChild(MatPageContentComponent),
        __metadata("design:type", MatPageContentComponent)
    ], MatPageLoaderComponent.prototype, "pageContent", void 0);
MatPageLoaderComponent.ɵfac = function MatPageLoaderComponent_Factory(t) { return new (t || MatPageLoaderComponent)(); };
MatPageLoaderComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: MatPageLoaderComponent, selectors: [["mat-page-loader"]], contentQueries: function MatPageLoaderComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, MatPageOutlookComponent, true);
        ɵngcc0.ɵɵcontentQuery(dirIndex, MatPageContentComponent, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.pageOutlook = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.pageContent = _t.first);
    } }, decls: 0, vars: 0, template: function MatPageLoaderComponent_Template(rf, ctx) { }, styles: [""] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(MatPageLoaderComponent, [{
        type: Component,
        args: [{
                selector: 'mat-page-loader',
                template: "",
                styles: [""]
            }]
    }], function () { return []; }, { pageOutlook: [{
            type: ContentChild,
            args: [MatPageOutlookComponent]
        }], pageContent: [{
            type: ContentChild,
            args: [MatPageContentComponent]
        }] }); })();
    return MatPageLoaderComponent;
}());

var MatPagesComponent = /** @class */ (function () {
    function MatPagesComponent(renderer) {
        this.renderer = renderer;
        this.done = new EventEmitter();
        this.pageChanged = new EventEmitter();
        // the mode of the sidenav
        this.navMode = 'side';
        this.index = 0;
    }
    MatPagesComponent.prototype._initSidenavMode = function () {
        if (window.innerWidth < 768) {
            this.navMode = 'over';
        }
    };
    MatPagesComponent.prototype.ngOnInit = function () {
        this._initSidenavMode();
    };
    MatPagesComponent.prototype.ngAfterContentInit = function () {
        this.totalPages = this.pages.length;
        this.renderer.setStyle(this.tabHeader.nativeElement.children[0], 'display', 'none');
    };
    MatPagesComponent.prototype.onResize = function (event) {
        if (event.target.innerWidth < 768) {
            this.navMode = 'over';
            this.sidenav.close();
        }
        if (event.target.innerWidth > 768) {
            this.navMode = 'side';
            this.sidenav.open();
        }
    };
    /**
     * Go to the previous page if it exists
     */
    MatPagesComponent.prototype.previous = function () {
        this.index = this.index === 0 ? 0 : --this.index;
        this.pageChanged.emit(this.index);
    };
    /**
     * Go to the next page if it exists
     */
    MatPagesComponent.prototype.next = function () {
        this.index = this.index === this.totalPages - 1 ? this.totalPages - 1 : ++this.index;
        this.pageChanged.emit(this.index);
        if (this.index === this.totalPages - 1) {
            this.done.emit();
        }
    };
    /**
     * Update the value of the index when a step has been selected.
     * Emit this value for pagechanged's subscribers.
     *
     */
    MatPagesComponent.prototype.onStepSelectionChange = function (event) {
        this.index = event.selectedIndex;
        this.pageChanged.emit(this.index);
    };
    MatPagesComponent.ctorParameters = function () { return [
        { type: Renderer2 }
    ]; };
    __decorate([
        ViewChild('sidenav'),
        __metadata("design:type", MatSidenav)
    ], MatPagesComponent.prototype, "sidenav", void 0);
    __decorate([
        ViewChild(MatTabGroup, { read: ElementRef, static: true }),
        __metadata("design:type", ElementRef)
    ], MatPagesComponent.prototype, "tabHeader", void 0);
    __decorate([
        ContentChildren(MatPageLoaderComponent),
        __metadata("design:type", QueryList)
    ], MatPagesComponent.prototype, "pages", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], MatPagesComponent.prototype, "title", void 0);
    __decorate([
        Output(),
        __metadata("design:type", EventEmitter)
    ], MatPagesComponent.prototype, "done", void 0);
    __decorate([
        Output(),
        __metadata("design:type", EventEmitter)
    ], MatPagesComponent.prototype, "pageChanged", void 0);
    __decorate([
        HostListener('window:resize', ['$event']),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", void 0)
    ], MatPagesComponent.prototype, "onResize", null);
    MatPagesComponent = __decorate([ __metadata("design:paramtypes", [Renderer2])
    ], MatPagesComponent);
MatPagesComponent.ɵfac = function MatPagesComponent_Factory(t) { return new (t || MatPagesComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Renderer2)); };
MatPagesComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: MatPagesComponent, selectors: [["mat-pages"]], contentQueries: function MatPagesComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, MatPageLoaderComponent, false);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.pages = _t);
    } }, viewQuery: function MatPagesComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c1, true);
        ɵngcc0.ɵɵstaticViewQuery(MatTabGroup, true, ElementRef);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.sidenav = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.tabHeader = _t.first);
    } }, hostBindings: function MatPagesComponent_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("resize", function MatPagesComponent_resize_HostBindingHandler($event) { return ctx.onResize($event); }, false, ɵngcc0.ɵɵresolveWindow);
    } }, inputs: { title: "title" }, outputs: { done: "done", pageChanged: "pageChanged" }, decls: 27, vars: 13, consts: [["opened", "true", 3, "mode", "fixedInViewport", "fixedTopGap", "fixedBottomGap"], ["sidenav", ""], ["color", "primary"], [3, "selectedIndex", "selectionChange"], [3, "label", 4, "ngFor", "ngForOf"], ["mat-icon-button", "", 3, "click"], [1, "demo-sidenav-content", "main-container"], [2, "min-height", "15vh", 3, "selectedIndex"], ["mat-fab", "", "color", "warn", 3, "click"], [4, "ngFor", "ngForOf"], [1, "navigation-buttons"], ["mat-fab", "", "color", "primary", 3, "disabled", "click"], [4, "ngIf", "ngIfThen", "ngIfElse"], ["nextAvailable", ""], ["finished", ""], [3, "label"], [3, "ngTemplateOutlet"], ["matTabContent", ""], ["mat-fab", "", "color", "primary", 3, "click"], ["mat-fab", "", "color", "accent", 3, "click"]], template: function MatPagesComponent_Template(rf, ctx) { if (rf & 1) {
        var _r18 = ɵngcc0.ɵɵgetCurrentView();
        ɵngcc0.ɵɵelementStart(0, "mat-sidenav-container");
        ɵngcc0.ɵɵelementStart(1, "mat-sidenav", 0, 1);
        ɵngcc0.ɵɵelement(3, "mat-toolbar", 2);
        ɵngcc0.ɵɵelementStart(4, "mat-vertical-stepper", 3);
        ɵngcc0.ɵɵlistener("selectionChange", function MatPagesComponent_Template_mat_vertical_stepper_selectionChange_4_listener($event) { return ctx.onStepSelectionChange($event); });
        ɵngcc0.ɵɵtemplate(5, MatPagesComponent_mat_step_5_Template, 2, 2, "mat-step", 4);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(6, "mat-sidenav-content");
        ɵngcc0.ɵɵelementStart(7, "mat-toolbar", 2);
        ɵngcc0.ɵɵelementStart(8, "button", 5);
        ɵngcc0.ɵɵlistener("click", function MatPagesComponent_Template_button_click_8_listener() { ɵngcc0.ɵɵrestoreView(_r18); var _r0 = ɵngcc0.ɵɵreference(2); return _r0.toggle(); });
        ɵngcc0.ɵɵelementStart(9, "mat-icon");
        ɵngcc0.ɵɵtext(10, "menu");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtext(11);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(12, "div", 6);
        ɵngcc0.ɵɵelementStart(13, "mat-tab-group", 7);
        ɵngcc0.ɵɵelementStart(14, "button", 8);
        ɵngcc0.ɵɵlistener("click", function MatPagesComponent_Template_button_click_14_listener() { return ctx.previous(); });
        ɵngcc0.ɵɵelementStart(15, "mat-icon");
        ɵngcc0.ɵɵtext(16, "location");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(17, MatPagesComponent_mat_tab_17_Template, 2, 0, "mat-tab", 9);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(18, "div", 10);
        ɵngcc0.ɵɵelementStart(19, "button", 11);
        ɵngcc0.ɵɵlistener("click", function MatPagesComponent_Template_button_click_19_listener() { return ctx.previous(); });
        ɵngcc0.ɵɵelementStart(20, "mat-icon");
        ɵngcc0.ɵɵtext(21, "chevron_left");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(22, MatPagesComponent_div_22_Template, 1, 0, "div", 12);
        ɵngcc0.ɵɵtemplate(23, MatPagesComponent_ng_template_23_Template, 3, 0, "ng-template", null, 13, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵtemplate(25, MatPagesComponent_ng_template_25_Template, 3, 0, "ng-template", null, 14, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        var _r4 = ɵngcc0.ɵɵreference(24);
        var _r6 = ɵngcc0.ɵɵreference(26);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("mode", ctx.navMode)("fixedInViewport", "false")("fixedTopGap", "false")("fixedBottomGap", "false");
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("selectedIndex", ctx.index);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.pages);
        ɵngcc0.ɵɵadvance(6);
        ɵngcc0.ɵɵtextInterpolate1(" ", ctx.title, " ");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("selectedIndex", ctx.index);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.pages);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("disabled", ctx.index == 0);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", ctx.index == ctx.totalPages - 1)("ngIfThen", _r6)("ngIfElse", _r4);
    } }, directives: [ɵngcc2.MatSidenavContainer, ɵngcc2.MatSidenav, ɵngcc3.MatToolbar, ɵngcc4.MatVerticalStepper, ɵngcc5.NgForOf, ɵngcc2.MatSidenavContent, ɵngcc6.MatButton, ɵngcc7.MatIcon, ɵngcc8.MatTabGroup, ɵngcc5.NgIf, ɵngcc4.MatStep, ɵngcc5.NgTemplateOutlet, ɵngcc8.MatTab, ɵngcc8.MatTabContent], styles: [".navigation-buttons[_ngcontent-%COMP%]{margin:1rem;place-content:stretch space-between;-webkit-box-align:stretch;align-items:stretch;-webkit-box-orient:horizontal;-webkit-box-direction:normal;flex-direction:row;box-sizing:border-box;display:-webkit-box;display:flex}"] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(MatPagesComponent, [{
        type: Component,
        args: [{
                selector: 'mat-pages',
                template: "<mat-sidenav-container>\n  <mat-sidenav #sidenav [mode]=\"navMode\" opened=\"true\"\n               [fixedInViewport]=\"'false'\" [fixedTopGap]=\"'false'\" [fixedBottomGap]=\"'false'\">\n\n    <mat-toolbar color=\"primary\">\n    </mat-toolbar>\n\n    <mat-vertical-stepper [selectedIndex]=\"index\" (selectionChange)=\"onStepSelectionChange($event)\">\n      <mat-step *ngFor=\"let page of pages;let i = index;\" [label]=\"page.pageOutlook.title\">\n        <ng-container [ngTemplateOutlet]=\"page.pageOutlook.content\"></ng-container>\n      </mat-step>\n    </mat-vertical-stepper>\n  </mat-sidenav>\n\n  <mat-sidenav-content>\n    <mat-toolbar color=\"primary\">\n      <button mat-icon-button (click)=\"sidenav.toggle()\">\n        <mat-icon>menu</mat-icon>\n      </button>\n      {{title}}\n    </mat-toolbar>\n    <div class=\"demo-sidenav-content main-container\">\n\n      <mat-tab-group [selectedIndex]=\"index\" style=\"min-height: 15vh;\" >\n        <button mat-fab color=\"warn\" (click)=\"previous()\">\n          <mat-icon>location</mat-icon>\n        </button>\n        <mat-tab *ngFor=\"let page of pages;let i = index;\">\n          <ng-template matTabContent>\n            <ng-container [ngTemplateOutlet]=\"page.pageContent.content\"></ng-container>\n          </ng-template>\n        </mat-tab>\n      </mat-tab-group>\n\n\n      <div class=\"navigation-buttons\">\n        <button mat-fab color=\"primary\" (click)=\"previous()\" [disabled]=\"index == 0\">\n          <mat-icon>chevron_left</mat-icon>\n        </button>\n        <div *ngIf=\"index == totalPages - 1;then finished else nextAvailable\">\n        </div>\n        <ng-template #nextAvailable>\n          <button mat-fab color=\"primary\" (click)=\"next()\">\n            <mat-icon>chevron_right</mat-icon>\n          </button>\n        </ng-template>\n        <ng-template #finished>\n          <button mat-fab color=\"accent\" (click)=\"done.next()\">\n            <mat-icon>done</mat-icon>\n          </button>\n        </ng-template>\n      </div>\n\n    </div>\n  </mat-sidenav-content>\n</mat-sidenav-container>\n",
                styles: [".navigation-buttons{margin:1rem;place-content:stretch space-between;-webkit-box-align:stretch;align-items:stretch;-webkit-box-orient:horizontal;-webkit-box-direction:normal;flex-direction:row;box-sizing:border-box;display:-webkit-box;display:flex}"]
            }]
    }], function () { return [{ type: ɵngcc0.Renderer2 }]; }, { done: [{
            type: Output
        }], pageChanged: [{
            type: Output
        }], onResize: [{
            type: HostListener,
            args: ['window:resize', ['$event']]
        }], sidenav: [{
            type: ViewChild,
            args: ['sidenav']
        }], tabHeader: [{
            type: ViewChild,
            args: [MatTabGroup, { read: ElementRef, static: true }]
        }], pages: [{
            type: ContentChildren,
            args: [MatPageLoaderComponent]
        }], title: [{
            type: Input
        }] }); })();
    return MatPagesComponent;
}());

var MatPagesModule = /** @class */ (function () {
    function MatPagesModule() {
    }
    MatPagesModule_1 = MatPagesModule;
    MatPagesModule.forRoot = function () {
        return {
            ngModule: MatPagesModule_1,
            providers: []
        };
    };
    var MatPagesModule_1;
MatPagesModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: MatPagesModule });
MatPagesModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function MatPagesModule_Factory(t) { return new (t || MatPagesModule)(); }, imports: [[
            CommonModule,
            MatToolbarModule,
            MatTabsModule,
            MatCardModule,
            MatSidenavModule,
            MatIconModule,
            MatStepperModule,
            MatButtonModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(MatPagesModule, { declarations: function () { return [MatPagesComponent,
        MatPageLoaderComponent,
        MatPageOutlookComponent,
        MatPageContentComponent]; }, imports: function () { return [CommonModule,
        MatToolbarModule,
        MatTabsModule,
        MatCardModule,
        MatSidenavModule,
        MatIconModule,
        MatStepperModule,
        MatButtonModule]; }, exports: function () { return [MatPagesComponent,
        MatPageLoaderComponent,
        MatPageOutlookComponent,
        MatPageContentComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(MatPagesModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    MatToolbarModule,
                    MatTabsModule,
                    MatCardModule,
                    MatSidenavModule,
                    MatIconModule,
                    MatStepperModule,
                    MatButtonModule
                ],
                exports: [
                    MatPagesComponent,
                    MatPageLoaderComponent,
                    MatPageOutlookComponent,
                    MatPageContentComponent
                ],
                declarations: [
                    MatPagesComponent,
                    MatPageLoaderComponent,
                    MatPageOutlookComponent,
                    MatPageContentComponent
                ]
            }]
    }], function () { return []; }, null); })();
    return MatPagesModule;
}());

/*
 * Public API Surface of pages
 */
// export * from './lib/pages.module';

/**
 * Generated bundle index. Do not edit.
 */

export { MatPageContentComponent, MatPageLoaderComponent, MatPageOutlookComponent, MatPagesComponent, MatPagesModule };
