// @alyle/ui/esm2015/snack-bar/snack-bar.module.js
import { NgModule } from '@angular/core';
import { LyOverlayModule } from '@alyle/ui';
import { LySnackBar } from './snack-bar';
import * as ɵngcc0 from '@angular/core';
export class LySnackBarModule {
}
LySnackBarModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: LySnackBarModule });
LySnackBarModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function LySnackBarModule_Factory(t) { return new (t || LySnackBarModule)(); }, imports: [[LyOverlayModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(LySnackBarModule, { declarations: function () { return [LySnackBar]; }, imports: function () { return [LyOverlayModule]; }, exports: function () { return [LySnackBar]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(LySnackBarModule, [{
        type: NgModule,
        args: [{
                imports: [LyOverlayModule],
                declarations: [LySnackBar],
                exports: [LySnackBar]
            }]
    }], null, null); })();
