// @alyle/ui/esm2015/themes/minima/deep-dark.js
import { Color } from '@alyle/ui/color';
import { MinimaDark } from './dark';
import { Injectable } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
const shadow = new Color(0, 0, 0, 1);
export class MinimaDeepDark extends MinimaDark {
    constructor() {
        super(...arguments);
        this.name = 'minima-deep-dark';
        this.background = {
            default: new Color(24, 24, 24),
            primary: {
                default: new Color(16, 16, 16),
                shadow
            },
            secondary: new Color(24, 24, 24),
            tertiary: new Color(32, 32, 32),
        };
        this.paper = {
            default: new Color(16, 16, 16),
            shadow
        };
        // field: LyFieldTheme = mergeThemes<LyFieldTheme, LyFieldTheme>(this.field, {
        //   root: _ => (className: string) => ``,
        //   appearance: {
        //     filled: _ => (className: string) => ``
        //   }
        // });
    }
}
MinimaDeepDark.ɵfac = function MinimaDeepDark_Factory(t) { return ɵMinimaDeepDark_BaseFactory(t || MinimaDeepDark); };
MinimaDeepDark.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: MinimaDeepDark, factory: MinimaDeepDark.ɵfac });
const ɵMinimaDeepDark_BaseFactory = ɵngcc0.ɵɵgetInheritedFactory(MinimaDeepDark);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(MinimaDeepDark, [{
        type: Injectable
    }], null, null); })();
