// @alyle/ui/esm2015/themes/minima/theme.js
import { NgModule, Directive } from '@angular/core';
import { LY_THEME_NAME, LyTheme2 } from '@alyle/ui';
import * as ɵngcc0 from '@angular/core';
export class ThemeMinimaLight {
}
ThemeMinimaLight.ɵfac = function ThemeMinimaLight_Factory(t) { return new (t || ThemeMinimaLight)(); };
ThemeMinimaLight.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThemeMinimaLight, selectors: [["", "ly-theme-minima-light", ""]], features: [ɵngcc0.ɵɵProvidersFeature([LyTheme2, { provide: LY_THEME_NAME, useValue: 'minima-light' }])] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThemeMinimaLight, [{
        type: Directive,
        args: [{
                selector: '[ly-theme-minima-light]',
                providers: [LyTheme2, { provide: LY_THEME_NAME, useValue: 'minima-light' }]
            }]
    }], null, null); })();
export class ThemeMinimaDark {
}
ThemeMinimaDark.ɵfac = function ThemeMinimaDark_Factory(t) { return new (t || ThemeMinimaDark)(); };
ThemeMinimaDark.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ThemeMinimaDark, selectors: [["", "ly-theme-minima-dark", ""]], features: [ɵngcc0.ɵɵProvidersFeature([LyTheme2, { provide: LY_THEME_NAME, useValue: 'minima-dark' }])] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThemeMinimaDark, [{
        type: Directive,
        args: [{
                selector: '[ly-theme-minima-dark]',
                providers: [LyTheme2, { provide: LY_THEME_NAME, useValue: 'minima-dark' }]
            }]
    }], null, null); })();
export class ThemeMinimaModule {
    constructor() {
        console.warn(`ThemeMinimaModule is deprecated.`);
    }
}
ThemeMinimaModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ThemeMinimaModule });
ThemeMinimaModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ThemeMinimaModule_Factory(t) { return new (t || ThemeMinimaModule)(); } });
ThemeMinimaModule.ctorParameters = () => [];
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ThemeMinimaModule, { declarations: [ThemeMinimaDark,
        ThemeMinimaLight], exports: [ThemeMinimaDark,
        ThemeMinimaLight] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ThemeMinimaModule, [{
        type: NgModule,
        args: [{
                declarations: [ThemeMinimaDark, ThemeMinimaLight],
                exports: [ThemeMinimaDark, ThemeMinimaLight]
            }]
    }], function () { return []; }, null); })();
