// @angular-material-extensions/pages/esm2015/lib/mat-page-content/mat-page-content.component.js
import { __decorate, __metadata } from "tslib";
import { Component, Input, TemplateRef, ViewChild } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/material/card';

function MatPageContentComponent_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "mat-card", 0);
    ɵngcc0.ɵɵelementStart(1, "mat-card-title");
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "mat-card-content");
    ɵngcc0.ɵɵprojection(4);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ctx_r0.title);
} }
const _c0 = ["*"];
let MatPageContentComponent = class MatPageContentComponent {
    constructor() {
    }
    ngOnInit() {
    }
};
MatPageContentComponent.ɵfac = function MatPageContentComponent_Factory(t) { return new (t || MatPageContentComponent)(); };
MatPageContentComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: MatPageContentComponent, selectors: [["mat-page-content"]], viewQuery: function MatPageContentComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(TemplateRef, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.content = _t.first);
    } }, inputs: { title: "title" }, ngContentSelectors: _c0, decls: 1, vars: 0, consts: [[2, "margin", "2rem"]], template: function MatPageContentComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵtemplate(0, MatPageContentComponent_ng_template_0_Template, 5, 1, "ng-template");
    } }, directives: [ɵngcc1.MatCard, ɵngcc1.MatCardTitle, ɵngcc1.MatCardContent], styles: [""] });
__decorate([
    ViewChild(TemplateRef),
    __metadata("design:type", TemplateRef)
], MatPageContentComponent.prototype, "content", void 0);
__decorate([
    Input(),
    __metadata("design:type", String)
], MatPageContentComponent.prototype, "title", void 0);
MatPageContentComponent = __decorate([ __metadata("design:paramtypes", [])
], MatPageContentComponent);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(MatPageContentComponent, [{
        type: Component,
        args: [{
                selector: 'mat-page-content',
                template: "<ng-template>\n  <mat-card style=\"margin: 2rem\">\n    <mat-card-title>{{title}}</mat-card-title>\n    <mat-card-content>\n      <ng-content></ng-content>\n    </mat-card-content>\n  </mat-card>\n</ng-template>\n",
                styles: [""]
            }]
    }], function () { return []; }, { content: [{
            type: ViewChild,
            args: [TemplateRef]
        }], title: [{
            type: Input
        }] }); })();
export { MatPageContentComponent };
