// @alyle/ui/esm2015/src/minimal/common.js
import { Directive, Input, ViewContainerRef, NgModule, ElementRef } from '@angular/core';
import { Platform } from '@angular/cdk/platform';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/cdk/platform';
export class NgTranscludeDirective {
    constructor(vcr, _platform) {
        this.vcr = vcr;
        this._platform = _platform;
    }
    set ngTransclude(templateRef) {
        if (templateRef && !this._ngTransclude) {
            if (this._platform.isBrowser && this._timeoutId != null) {
                window.clearTimeout(this._timeoutId);
                this._timeoutId = null;
                this.vcr.clear();
            }
            this._ngTransclude = templateRef;
            this.vcr.createEmbeddedView(templateRef);
        }
        else if (this._ngTransclude && !templateRef) {
            this._ngTransclude = null;
            this.clear();
        }
    }
    get getNgTransclude() {
        return this._ngTransclude;
    }
    clear() {
        if (this._platform.isBrowser && this.timeout) {
            this._timeoutId = window.setTimeout(() => {
                this.vcr.clear();
                this._timeoutId = null;
            }, this.timeout);
        }
        else {
            this.vcr.clear();
        }
    }
    ngOnDestroy() {
        if (this._platform.isBrowser) {
            window.clearTimeout(this._timeoutId);
        }
        this.vcr.clear();
    }
}
NgTranscludeDirective.ɵfac = function NgTranscludeDirective_Factory(t) { return new (t || NgTranscludeDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ViewContainerRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.Platform)); };
NgTranscludeDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: NgTranscludeDirective, selectors: [["", "ngTransclude", ""]], inputs: { ngTransclude: "ngTransclude", timeout: "timeout" }, exportAs: ["ngTransclude"] });
NgTranscludeDirective.ctorParameters = () => [
    { type: ViewContainerRef },
    { type: Platform }
];
NgTranscludeDirective.propDecorators = {
    timeout: [{ type: Input }],
    ngTransclude: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgTranscludeDirective, [{
        type: Directive,
        args: [{
                selector: '[ngTransclude]',
                exportAs: 'ngTransclude'
            }]
    }], function () { return [{ type: ɵngcc0.ViewContainerRef }, { type: ɵngcc1.Platform }]; }, { ngTransclude: [{
            type: Input
        }], timeout: [{
            type: Input
        }] }); })();
export class NgTranscludeModule {
}
NgTranscludeModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NgTranscludeModule });
NgTranscludeModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NgTranscludeModule_Factory(t) { return new (t || NgTranscludeModule)(); } });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NgTranscludeModule, { declarations: [NgTranscludeDirective], exports: [NgTranscludeDirective] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgTranscludeModule, [{
        type: NgModule,
        args: [{
                exports: [NgTranscludeDirective],
                declarations: [NgTranscludeDirective]
            }]
    }], null, null); })();
/**
 * @ignore
 */
export function getNativeElement(element) {
    return element instanceof ElementRef ? element.nativeElement : element;
}
