// @alyle/ui/esm2015/drawer/drawer.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { LyDrawer, LyDrawerContainer, LyDrawerContent } from './drawer';
import { LyCommonModule } from '@alyle/ui';
import * as ɵngcc0 from '@angular/core';
export class LyDrawerModule {
}
LyDrawerModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: LyDrawerModule });
LyDrawerModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function LyDrawerModule_Factory(t) { return new (t || LyDrawerModule)(); }, imports: [[
            CommonModule,
            LyCommonModule
        ],
        LyCommonModule] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(LyDrawerModule, { declarations: function () { return [LyDrawer, LyDrawerContainer, LyDrawerContent]; }, imports: function () { return [CommonModule,
        LyCommonModule]; }, exports: function () { return [LyDrawer,
        LyDrawerContainer,
        LyDrawerContent,
        LyCommonModule]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(LyDrawerModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    LyCommonModule
                ],
                exports: [
                    LyDrawer,
                    LyDrawerContainer,
                    LyDrawerContent,
                    LyCommonModule
                ],
                declarations: [LyDrawer, LyDrawerContainer, LyDrawerContent]
            }]
    }], null, null); })();
