// @alyle/ui/esm2015/src/ripple/ripple.service.js
import { Injectable } from '@angular/core';
import { LY_COMMON_STYLES_DEPRECATED } from '../styles/core-styles';
import { LyTheme2 } from '../theme/theme2.service';
import * as i0 from "@angular/core";
import * as i1 from "../theme/theme2.service";
import * as ɵngcc0 from '@angular/core';
export const styles = (theme) => ({
    rippleContainer: {
        position: 'absolute',
        width: '2px',
        height: '2px',
        background: 'currentColor',
        opacity: '.2',
        borderRadius: '50%',
        transform: 'scale(0)',
        transition: `opacity ${theme.ripple.transition.opacity},transform ${theme.ripple.transition.transform}`,
        pointerEvents: 'none'
    },
    container: Object.assign(Object.assign({}, LY_COMMON_STYLES_DEPRECATED.fill), { overflow: 'hidden', pointerEvents: 'none', borderRadius: 'inherit' })
});
export class LyRippleService {
    constructor(theme) {
        this.theme = theme;
        this.classes = this.theme.addStyleSheet(styles);
    }
}
LyRippleService.ɵfac = function LyRippleService_Factory(t) { return new (t || LyRippleService)(ɵngcc0.ɵɵinject(LyTheme2)); };
LyRippleService.ɵprov = i0.ɵɵdefineInjectable({ factory: function LyRippleService_Factory() { return new LyRippleService(i0.ɵɵinject(i1.LyTheme2)); }, token: LyRippleService, providedIn: "root" });
LyRippleService.ctorParameters = () => [
    { type: LyTheme2 }
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(LyRippleService, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: LyTheme2 }]; }, null); })();
