// @alyle/ui/esm2015/list/list.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { LyCommonModule } from '@alyle/ui';
import { LyList, LyListItem, LyListIcon, LyLine } from './list';
import * as ɵngcc0 from '@angular/core';
export class LyListModule {
}
LyListModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: LyListModule });
LyListModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function LyListModule_Factory(t) { return new (t || LyListModule)(); }, imports: [[
            CommonModule
        ],
        LyCommonModule] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(LyListModule, { declarations: function () { return [LyList, LyListItem, LyListIcon, LyLine]; }, imports: function () { return [CommonModule]; }, exports: function () { return [LyCommonModule, LyList, LyListItem, LyListIcon, LyLine]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(LyListModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule
                ],
                declarations: [LyList, LyListItem, LyListIcon, LyLine],
                exports: [LyCommonModule, LyList, LyListItem, LyListIcon, LyLine]
            }]
    }], null, null); })();
