// @angular-material-extensions/pages/esm2015/lib/mat-page-loader/mat-page-loader.component.js
import { __decorate, __metadata } from "tslib";
import { Component, ContentChild } from '@angular/core';
import { MatPageOutlookComponent } from '../mat-page-outlook/mat-page-outlook.component';
import { MatPageContentComponent } from '../mat-page-content/mat-page-content.component';
import * as ɵngcc0 from '@angular/core';
let MatPageLoaderComponent = class MatPageLoaderComponent {
};
MatPageLoaderComponent.ɵfac = function MatPageLoaderComponent_Factory(t) { return new (t || MatPageLoaderComponent)(); };
MatPageLoaderComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: MatPageLoaderComponent, selectors: [["mat-page-loader"]], contentQueries: function MatPageLoaderComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, MatPageOutlookComponent, true);
        ɵngcc0.ɵɵcontentQuery(dirIndex, MatPageContentComponent, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.pageOutlook = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.pageContent = _t.first);
    } }, decls: 0, vars: 0, template: function MatPageLoaderComponent_Template(rf, ctx) { }, styles: [""] });
__decorate([
    ContentChild(MatPageOutlookComponent),
    __metadata("design:type", MatPageOutlookComponent)
], MatPageLoaderComponent.prototype, "pageOutlook", void 0);
__decorate([
    ContentChild(MatPageContentComponent),
    __metadata("design:type", MatPageContentComponent)
], MatPageLoaderComponent.prototype, "pageContent", void 0);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(MatPageLoaderComponent, [{
        type: Component,
        args: [{
                selector: 'mat-page-loader',
                template: "",
                styles: [""]
            }]
    }], null, { pageOutlook: [{
            type: ContentChild,
            args: [MatPageOutlookComponent]
        }], pageContent: [{
            type: ContentChild,
            args: [MatPageContentComponent]
        }] }); })();
export { MatPageLoaderComponent };
