// @material-extended/mde/esm5/lib/popover/popover-module.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { OverlayModule } from '@angular/cdk/overlay';
import { MdePopover } from './popover';
import { MdePopoverTrigger } from './popover-trigger';
import { MdePopoverTarget } from './popover-target';
import { A11yModule } from '@angular/cdk/a11y';
import * as ɵngcc0 from '@angular/core';
var MdePopoverModule = /** @class */ (function () {
    function MdePopoverModule() {
    }
MdePopoverModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: MdePopoverModule });
MdePopoverModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function MdePopoverModule_Factory(t) { return new (t || MdePopoverModule)(); }, imports: [[
            OverlayModule,
            CommonModule,
            A11yModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(MdePopoverModule, { declarations: function () { return [MdePopover, MdePopoverTrigger, MdePopoverTarget]; }, imports: function () { return [OverlayModule,
        CommonModule,
        A11yModule]; }, exports: function () { return [MdePopover, MdePopoverTrigger, MdePopoverTarget]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(MdePopoverModule, [{
        type: NgModule,
        args: [{
                imports: [
                    OverlayModule,
                    CommonModule,
                    A11yModule
                ],
                exports: [MdePopover, MdePopoverTrigger, MdePopoverTarget],
                declarations: [MdePopover, MdePopoverTrigger, MdePopoverTarget]
            }]
    }], function () { return []; }, null); })();
    return MdePopoverModule;
}());
export { MdePopoverModule };
