// @alyle/ui/esm2015/checkbox/checkbox.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { LyCommonModule } from '@alyle/ui';
import { LyCheckbox } from './checkbox';
import * as ɵngcc0 from '@angular/core';
export class LyCheckboxModule {
}
LyCheckboxModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: LyCheckboxModule });
LyCheckboxModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function LyCheckboxModule_Factory(t) { return new (t || LyCheckboxModule)(); }, imports: [[
            CommonModule,
            LyCommonModule
        ],
        LyCommonModule] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(LyCheckboxModule, { declarations: function () { return [LyCheckbox]; }, imports: function () { return [CommonModule,
        LyCommonModule]; }, exports: function () { return [LyCommonModule,
        LyCheckbox]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(LyCheckboxModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    LyCheckbox
                ],
                imports: [
                    CommonModule,
                    LyCommonModule
                ],
                exports: [
                    LyCommonModule,
                    LyCheckbox
                ]
            }]
    }], null, null); })();
