// @alyle/ui/esm2015/expansion/expansion.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { NgTranscludeModule, LyExpansionIconModule } from '@alyle/ui';
import { LyAccordion } from './accordion';
import { LyExpansionPanel } from './expansion-panel';
import { LyExpansionPanelHeader } from './expansion-panel-header';
import { LyExpansionPanelContent } from './expansion-panel-content';
import { LyExpansionPanelTitle } from './expansion-panel-title';
import { LyExpansionPanelDescription } from './expansion-panel-description';
import { LyExpansionPanelAction } from './expansion-panel-action-row';
import * as ɵngcc0 from '@angular/core';
export class LyExpansionModule {
}
LyExpansionModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: LyExpansionModule });
LyExpansionModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function LyExpansionModule_Factory(t) { return new (t || LyExpansionModule)(); }, imports: [[
            CommonModule,
            LyExpansionIconModule,
            NgTranscludeModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(LyExpansionModule, { declarations: function () { return [LyAccordion,
        LyExpansionPanel,
        LyExpansionPanelHeader,
        LyExpansionPanelContent,
        LyExpansionPanelTitle,
        LyExpansionPanelDescription,
        LyExpansionPanelAction]; }, imports: function () { return [CommonModule,
        LyExpansionIconModule,
        NgTranscludeModule]; }, exports: function () { return [LyAccordion,
        LyExpansionPanel,
        LyExpansionPanelHeader,
        LyExpansionPanelContent,
        LyExpansionPanelTitle,
        LyExpansionPanelDescription,
        LyExpansionPanelAction]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(LyExpansionModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    LyAccordion,
                    LyExpansionPanel,
                    LyExpansionPanelHeader,
                    LyExpansionPanelContent,
                    LyExpansionPanelTitle,
                    LyExpansionPanelDescription,
                    LyExpansionPanelAction
                ],
                imports: [
                    CommonModule,
                    LyExpansionIconModule,
                    NgTranscludeModule
                ],
                exports: [
                    LyAccordion,
                    LyExpansionPanel,
                    LyExpansionPanelHeader,
                    LyExpansionPanelContent,
                    LyExpansionPanelTitle,
                    LyExpansionPanelDescription,
                    LyExpansionPanelAction
                ]
            }]
    }], null, null); })();
