// @alyle/ui/esm2015/expansion/expansion-panel-header.js
import { Component, ElementRef, Renderer2, Inject } from '@angular/core';
import { LyAccordion } from './accordion';
import { LyExpansionPanel } from './expansion-panel';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from '@alyle/ui';

function LyExpansionPanelHeader_ly_expansion_icon_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "ly-expansion-icon", 2);
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("up", ctx_r0._expansionPanel.expanded);
} }
const _c0 = [[["ly-panel-title"]], [["ly-panel-description"]], "*"];
const _c1 = ["ly-panel-title", "ly-panel-description", "*"];
export class LyExpansionPanelHeader {
    constructor(el, renderer, _accordion, _expansionPanel) {
        this._accordion = _accordion;
        this._expansionPanel = _expansionPanel;
        /** @docs-private */
        this.classes = this._accordion.classes;
        renderer.addClass(el.nativeElement, this._accordion.classes.panelHeader);
    }
}
LyExpansionPanelHeader.ɵfac = function LyExpansionPanelHeader_Factory(t) { return new (t || LyExpansionPanelHeader)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Renderer2), ɵngcc0.ɵɵdirectiveInject(LyAccordion), ɵngcc0.ɵɵdirectiveInject(LyExpansionPanel)); };
LyExpansionPanelHeader.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: LyExpansionPanelHeader, selectors: [["ly-expansion-panel-header"]], hostBindings: function LyExpansionPanelHeader_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("click", function LyExpansionPanelHeader_click_HostBindingHandler() { return ctx._expansionPanel.toggle(); });
    } }, ngContentSelectors: _c1, decls: 5, vars: 2, consts: [[3, "className"], [3, "up", 4, "ngIf"], [3, "up"]], template: function LyExpansionPanelHeader_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef(_c0);
        ɵngcc0.ɵɵelementStart(0, "span", 0);
        ɵngcc0.ɵɵprojection(1);
        ɵngcc0.ɵɵprojection(2, 1);
        ɵngcc0.ɵɵprojection(3, 2);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(4, LyExpansionPanelHeader_ly_expansion_icon_4_Template, 1, 1, "ly-expansion-icon", 1);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("className", ctx.classes.panelHeaderContent);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("ngIf", !ctx._expansionPanel.disabled && ctx._expansionPanel.hasToggle);
    } }, directives: [ɵngcc1.NgIf, ɵngcc2.LyExpansionIcon], encapsulation: 2 });
LyExpansionPanelHeader.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 },
    { type: LyAccordion, decorators: [{ type: Inject, args: [LyAccordion,] }] },
    { type: LyExpansionPanel, decorators: [{ type: Inject, args: [LyExpansionPanel,] }] }
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(LyExpansionPanelHeader, [{
        type: Component,
        args: [{
                selector: 'ly-expansion-panel-header',
                template: "<span [className]=\"classes.panelHeaderContent\">\n  <ng-content select=\"ly-panel-title\"></ng-content>\n  <ng-content select=\"ly-panel-description\"></ng-content>\n  <ng-content></ng-content>\n</span>\n<ly-expansion-icon\n  *ngIf=\"!_expansionPanel.disabled && _expansionPanel.hasToggle\"\n  [up]=\"_expansionPanel.expanded\"\n></ly-expansion-icon>",
                host: {
                    '(click)': '_expansionPanel.toggle()'
                }
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ɵngcc0.Renderer2 }, { type: LyAccordion, decorators: [{
                type: Inject,
                args: [LyAccordion]
            }] }, { type: LyExpansionPanel, decorators: [{
                type: Inject,
                args: [LyExpansionPanel]
            }] }]; }, null); })();
