// @alyle/ui/esm2015/responsive/responsive.module.js
import { NgModule } from '@angular/core';
import { MediaDirective } from './media.directive';
import { LyCommonModule } from '@alyle/ui';
/**
 * @deprecated use instead `[lyDisplay]` or `[lyStyle]`
 */
import * as ɵngcc0 from '@angular/core';
export class ResponsiveModule {
}
ResponsiveModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ResponsiveModule });
ResponsiveModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ResponsiveModule_Factory(t) { return new (t || ResponsiveModule)(); }, imports: [LyCommonModule] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ResponsiveModule, { declarations: function () { return [MediaDirective]; }, exports: function () { return [MediaDirective, LyCommonModule]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ResponsiveModule, [{
        type: NgModule,
        args: [{
                declarations: [MediaDirective],
                exports: [MediaDirective, LyCommonModule]
            }]
    }], null, null); })();
