// @angular-material-extensions/pages/bundles/angular-material-extensions-pages.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('@angular/material/card'),require('@angular/material/sidenav'),require('@angular/material/toolbar'),require('@angular/material/stepper'),require('@angular/common'),require('@angular/material/button'),require('@angular/material/icon'),require('@angular/material/tabs'),exports, require('@angular/core'), require('@angular/material/sidenav'), require('@angular/material/tabs'), require('@angular/common'), require('@angular/material/button'), require('@angular/material/card'), require('@angular/material/icon'), require('@angular/material/stepper'), require('@angular/material/toolbar')) :
    typeof define === 'function' && define.amd ? define('@angular-material-extensions/pages', ['@angular/core','@angular/material/card','@angular/material/sidenav','@angular/material/toolbar','@angular/material/stepper','@angular/common','@angular/material/button','@angular/material/icon','@angular/material/tabs','exports', '@angular/core', '@angular/material/sidenav', '@angular/material/tabs', '@angular/common', '@angular/material/button', '@angular/material/card', '@angular/material/icon', '@angular/material/stepper', '@angular/material/toolbar'], factory) :
    (global = global || self, factory(global.ng.core,global.ng.material.card,global.ng.material.sidenav,global.ng.material.toolbar,global.ng.material.stepper,global.ng.common,global.ng.material.button,global.ng.material.icon,global.ng.material.tabs,(global['angular-material-extensions'] = global['angular-material-extensions'] || {}, global['angular-material-extensions'].pages = {}), global.ng.core, global.ng.material.sidenav, global.ng.material.tabs, global.ng.common, global.ng.material.button, global.ng.material.card, global.ng.material.icon, global.ng.material.stepper, global.ng.material.toolbar));
}(this, (function (ɵngcc0,ɵngcc1,ɵngcc2,ɵngcc3,ɵngcc4,ɵngcc5,ɵngcc6,ɵngcc7,ɵngcc8,exports, core, sidenav, tabs, common, button, card, icon, stepper, toolbar) { 
function MatPageContentComponent_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "mat-card", 0);
    ɵngcc0.ɵɵelementStart(1, "mat-card-title");
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "mat-card-content");
    ɵngcc0.ɵɵprojection(4);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ctx_r0.title);
} }
var _c0 = ["*"];
function MatPageOutlookComponent_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵprojection(0);
} }
var _c1 = ["sidenav"];
function MatPagesComponent_mat_step_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "mat-step", 15);
    ɵngcc0.ɵɵelementContainer(1, 16);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var page_r8 = ctx.$implicit;
    ɵngcc0.ɵɵproperty("label", page_r8.pageOutlook.title);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", page_r8.pageOutlook.content);
} }
function MatPagesComponent_mat_tab_17_ng_template_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0, 16);
} if (rf & 2) {
    var page_r10 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", page_r10.pageContent.content);
} }
function MatPagesComponent_mat_tab_17_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "mat-tab");
    ɵngcc0.ɵɵtemplate(1, MatPagesComponent_mat_tab_17_ng_template_1_Template, 1, 1, "ng-template", 17);
    ɵngcc0.ɵɵelementEnd();
} }
function MatPagesComponent_div_22_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "div");
} }
function MatPagesComponent_ng_template_23_Template(rf, ctx) { if (rf & 1) {
    var _r15 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 18);
    ɵngcc0.ɵɵlistener("click", function MatPagesComponent_ng_template_23_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r15); var ctx_r14 = ɵngcc0.ɵɵnextContext(); return ctx_r14.next(); });
    ɵngcc0.ɵɵelementStart(1, "mat-icon");
    ɵngcc0.ɵɵtext(2, "chevron_right");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
function MatPagesComponent_ng_template_25_Template(rf, ctx) { if (rf & 1) {
    var _r17 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 19);
    ɵngcc0.ɵɵlistener("click", function MatPagesComponent_ng_template_25_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r17); var ctx_r16 = ɵngcc0.ɵɵnextContext(); return ctx_r16.done.next(); });
    ɵngcc0.ɵɵelementStart(1, "mat-icon");
    ɵngcc0.ɵɵtext(2, "done");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (_) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    function __exportStar(m, exports) {
        for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
    }

    function __values(o) {
        var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
        if (m) return m.call(o);
        return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    };

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
        result.default = mod;
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    var MatPageContentComponent = /** @class */ (function () {
        function MatPageContentComponent() {
        }
        MatPageContentComponent.prototype.ngOnInit = function () {
        };
        __decorate([
            core.ViewChild(core.TemplateRef),
            __metadata("design:type", core.TemplateRef)
        ], MatPageContentComponent.prototype, "content", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", String)
        ], MatPageContentComponent.prototype, "title", void 0);
        MatPageContentComponent = __decorate([ __metadata("design:paramtypes", [])
        ], MatPageContentComponent);
MatPageContentComponent.ɵfac = function MatPageContentComponent_Factory(t) { return new (t || MatPageContentComponent)(); };
MatPageContentComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: MatPageContentComponent, selectors: [["mat-page-content"]], viewQuery: function MatPageContentComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(core.TemplateRef, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.content = _t.first);
    } }, inputs: { title: "title" }, ngContentSelectors: _c0, decls: 1, vars: 0, consts: [[2, "margin", "2rem"]], template: function MatPageContentComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵtemplate(0, MatPageContentComponent_ng_template_0_Template, 5, 1, "ng-template");
    } }, directives: [ɵngcc1.MatCard, ɵngcc1.MatCardTitle, ɵngcc1.MatCardContent], styles: [""] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(MatPageContentComponent, [{
        type: core.Component,
        args: [{
                selector: 'mat-page-content',
                template: "<ng-template>\n  <mat-card style=\"margin: 2rem\">\n    <mat-card-title>{{title}}</mat-card-title>\n    <mat-card-content>\n      <ng-content></ng-content>\n    </mat-card-content>\n  </mat-card>\n</ng-template>\n",
                styles: [""]
            }]
    }], function () { return []; }, { content: [{
            type: core.ViewChild,
            args: [core.TemplateRef]
        }], title: [{
            type: core.Input
        }] }); })();
        return MatPageContentComponent;
    }());

    var MatPageOutlookComponent = /** @class */ (function () {
        function MatPageOutlookComponent() {
        }
        MatPageOutlookComponent.prototype.ngOnInit = function () {
        };
        __decorate([
            core.ViewChild(core.TemplateRef),
            __metadata("design:type", core.TemplateRef)
        ], MatPageOutlookComponent.prototype, "content", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", String)
        ], MatPageOutlookComponent.prototype, "title", void 0);
        MatPageOutlookComponent = __decorate([ __metadata("design:paramtypes", [])
        ], MatPageOutlookComponent);
MatPageOutlookComponent.ɵfac = function MatPageOutlookComponent_Factory(t) { return new (t || MatPageOutlookComponent)(); };
MatPageOutlookComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: MatPageOutlookComponent, selectors: [["mat-page-outlook"]], viewQuery: function MatPageOutlookComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(core.TemplateRef, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.content = _t.first);
    } }, inputs: { title: "title" }, ngContentSelectors: _c0, decls: 1, vars: 0, template: function MatPageOutlookComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵtemplate(0, MatPageOutlookComponent_ng_template_0_Template, 1, 0, "ng-template");
    } }, styles: [""], encapsulation: 2, changeDetection: 0 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(MatPageOutlookComponent, [{
        type: core.Component,
        args: [{
                selector: 'mat-page-outlook',
                template: "<ng-template>\n  <ng-content></ng-content>\n</ng-template>\n",
                encapsulation: core.ViewEncapsulation.None,
                changeDetection: core.ChangeDetectionStrategy.OnPush,
                styles: [""]
            }]
    }], function () { return []; }, { content: [{
            type: core.ViewChild,
            args: [core.TemplateRef]
        }], title: [{
            type: core.Input
        }] }); })();
        return MatPageOutlookComponent;
    }());

    var MatPageLoaderComponent = /** @class */ (function () {
        function MatPageLoaderComponent() {
        }
        __decorate([
            core.ContentChild(MatPageOutlookComponent),
            __metadata("design:type", MatPageOutlookComponent)
        ], MatPageLoaderComponent.prototype, "pageOutlook", void 0);
        __decorate([
            core.ContentChild(MatPageContentComponent),
            __metadata("design:type", MatPageContentComponent)
        ], MatPageLoaderComponent.prototype, "pageContent", void 0);
MatPageLoaderComponent.ɵfac = function MatPageLoaderComponent_Factory(t) { return new (t || MatPageLoaderComponent)(); };
MatPageLoaderComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: MatPageLoaderComponent, selectors: [["mat-page-loader"]], contentQueries: function MatPageLoaderComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, MatPageOutlookComponent, true);
        ɵngcc0.ɵɵcontentQuery(dirIndex, MatPageContentComponent, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.pageOutlook = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.pageContent = _t.first);
    } }, decls: 0, vars: 0, template: function MatPageLoaderComponent_Template(rf, ctx) { }, styles: [""] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(MatPageLoaderComponent, [{
        type: core.Component,
        args: [{
                selector: 'mat-page-loader',
                template: "",
                styles: [""]
            }]
    }], function () { return []; }, { pageOutlook: [{
            type: core.ContentChild,
            args: [MatPageOutlookComponent]
        }], pageContent: [{
            type: core.ContentChild,
            args: [MatPageContentComponent]
        }] }); })();
        return MatPageLoaderComponent;
    }());

    var MatPagesComponent = /** @class */ (function () {
        function MatPagesComponent(renderer) {
            this.renderer = renderer;
            this.done = new core.EventEmitter();
            this.pageChanged = new core.EventEmitter();
            // the mode of the sidenav
            this.navMode = 'side';
            this.index = 0;
        }
        MatPagesComponent.prototype._initSidenavMode = function () {
            if (window.innerWidth < 768) {
                this.navMode = 'over';
            }
        };
        MatPagesComponent.prototype.ngOnInit = function () {
            this._initSidenavMode();
        };
        MatPagesComponent.prototype.ngAfterContentInit = function () {
            this.totalPages = this.pages.length;
            this.renderer.setStyle(this.tabHeader.nativeElement.children[0], 'display', 'none');
        };
        MatPagesComponent.prototype.onResize = function (event) {
            if (event.target.innerWidth < 768) {
                this.navMode = 'over';
                this.sidenav.close();
            }
            if (event.target.innerWidth > 768) {
                this.navMode = 'side';
                this.sidenav.open();
            }
        };
        /**
         * Go to the previous page if it exists
         */
        MatPagesComponent.prototype.previous = function () {
            this.index = this.index === 0 ? 0 : --this.index;
            this.pageChanged.emit(this.index);
        };
        /**
         * Go to the next page if it exists
         */
        MatPagesComponent.prototype.next = function () {
            this.index = this.index === this.totalPages - 1 ? this.totalPages - 1 : ++this.index;
            this.pageChanged.emit(this.index);
            if (this.index === this.totalPages - 1) {
                this.done.emit();
            }
        };
        /**
         * Update the value of the index when a step has been selected.
         * Emit this value for pagechanged's subscribers.
         *
         */
        MatPagesComponent.prototype.onStepSelectionChange = function (event) {
            this.index = event.selectedIndex;
            this.pageChanged.emit(this.index);
        };
        MatPagesComponent.ctorParameters = function () { return [
            { type: core.Renderer2 }
        ]; };
        __decorate([
            core.ViewChild('sidenav'),
            __metadata("design:type", sidenav.MatSidenav)
        ], MatPagesComponent.prototype, "sidenav", void 0);
        __decorate([
            core.ViewChild(tabs.MatTabGroup, { read: core.ElementRef, static: true }),
            __metadata("design:type", core.ElementRef)
        ], MatPagesComponent.prototype, "tabHeader", void 0);
        __decorate([
            core.ContentChildren(MatPageLoaderComponent),
            __metadata("design:type", core.QueryList)
        ], MatPagesComponent.prototype, "pages", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", String)
        ], MatPagesComponent.prototype, "title", void 0);
        __decorate([
            core.Output(),
            __metadata("design:type", core.EventEmitter)
        ], MatPagesComponent.prototype, "done", void 0);
        __decorate([
            core.Output(),
            __metadata("design:type", core.EventEmitter)
        ], MatPagesComponent.prototype, "pageChanged", void 0);
        __decorate([
            core.HostListener('window:resize', ['$event']),
            __metadata("design:type", Function),
            __metadata("design:paramtypes", [Object]),
            __metadata("design:returntype", void 0)
        ], MatPagesComponent.prototype, "onResize", null);
        MatPagesComponent = __decorate([ __metadata("design:paramtypes", [core.Renderer2])
        ], MatPagesComponent);
MatPagesComponent.ɵfac = function MatPagesComponent_Factory(t) { return new (t || MatPagesComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Renderer2)); };
MatPagesComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: MatPagesComponent, selectors: [["mat-pages"]], contentQueries: function MatPagesComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, MatPageLoaderComponent, false);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.pages = _t);
    } }, viewQuery: function MatPagesComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c1, true);
        ɵngcc0.ɵɵstaticViewQuery(tabs.MatTabGroup, true, core.ElementRef);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.sidenav = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.tabHeader = _t.first);
    } }, hostBindings: function MatPagesComponent_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("resize", function MatPagesComponent_resize_HostBindingHandler($event) { return ctx.onResize($event); }, false, ɵngcc0.ɵɵresolveWindow);
    } }, inputs: { title: "title" }, outputs: { done: "done", pageChanged: "pageChanged" }, decls: 27, vars: 13, consts: [["opened", "true", 3, "mode", "fixedInViewport", "fixedTopGap", "fixedBottomGap"], ["sidenav", ""], ["color", "primary"], [3, "selectedIndex", "selectionChange"], [3, "label", 4, "ngFor", "ngForOf"], ["mat-icon-button", "", 3, "click"], [1, "demo-sidenav-content", "main-container"], [2, "min-height", "15vh", 3, "selectedIndex"], ["mat-fab", "", "color", "warn", 3, "click"], [4, "ngFor", "ngForOf"], [1, "navigation-buttons"], ["mat-fab", "", "color", "primary", 3, "disabled", "click"], [4, "ngIf", "ngIfThen", "ngIfElse"], ["nextAvailable", ""], ["finished", ""], [3, "label"], [3, "ngTemplateOutlet"], ["matTabContent", ""], ["mat-fab", "", "color", "primary", 3, "click"], ["mat-fab", "", "color", "accent", 3, "click"]], template: function MatPagesComponent_Template(rf, ctx) { if (rf & 1) {
        var _r18 = ɵngcc0.ɵɵgetCurrentView();
        ɵngcc0.ɵɵelementStart(0, "mat-sidenav-container");
        ɵngcc0.ɵɵelementStart(1, "mat-sidenav", 0, 1);
        ɵngcc0.ɵɵelement(3, "mat-toolbar", 2);
        ɵngcc0.ɵɵelementStart(4, "mat-vertical-stepper", 3);
        ɵngcc0.ɵɵlistener("selectionChange", function MatPagesComponent_Template_mat_vertical_stepper_selectionChange_4_listener($event) { return ctx.onStepSelectionChange($event); });
        ɵngcc0.ɵɵtemplate(5, MatPagesComponent_mat_step_5_Template, 2, 2, "mat-step", 4);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(6, "mat-sidenav-content");
        ɵngcc0.ɵɵelementStart(7, "mat-toolbar", 2);
        ɵngcc0.ɵɵelementStart(8, "button", 5);
        ɵngcc0.ɵɵlistener("click", function MatPagesComponent_Template_button_click_8_listener() { ɵngcc0.ɵɵrestoreView(_r18); var _r0 = ɵngcc0.ɵɵreference(2); return _r0.toggle(); });
        ɵngcc0.ɵɵelementStart(9, "mat-icon");
        ɵngcc0.ɵɵtext(10, "menu");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtext(11);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(12, "div", 6);
        ɵngcc0.ɵɵelementStart(13, "mat-tab-group", 7);
        ɵngcc0.ɵɵelementStart(14, "button", 8);
        ɵngcc0.ɵɵlistener("click", function MatPagesComponent_Template_button_click_14_listener() { return ctx.previous(); });
        ɵngcc0.ɵɵelementStart(15, "mat-icon");
        ɵngcc0.ɵɵtext(16, "location");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(17, MatPagesComponent_mat_tab_17_Template, 2, 0, "mat-tab", 9);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(18, "div", 10);
        ɵngcc0.ɵɵelementStart(19, "button", 11);
        ɵngcc0.ɵɵlistener("click", function MatPagesComponent_Template_button_click_19_listener() { return ctx.previous(); });
        ɵngcc0.ɵɵelementStart(20, "mat-icon");
        ɵngcc0.ɵɵtext(21, "chevron_left");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(22, MatPagesComponent_div_22_Template, 1, 0, "div", 12);
        ɵngcc0.ɵɵtemplate(23, MatPagesComponent_ng_template_23_Template, 3, 0, "ng-template", null, 13, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵtemplate(25, MatPagesComponent_ng_template_25_Template, 3, 0, "ng-template", null, 14, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        var _r4 = ɵngcc0.ɵɵreference(24);
        var _r6 = ɵngcc0.ɵɵreference(26);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("mode", ctx.navMode)("fixedInViewport", "false")("fixedTopGap", "false")("fixedBottomGap", "false");
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("selectedIndex", ctx.index);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.pages);
        ɵngcc0.ɵɵadvance(6);
        ɵngcc0.ɵɵtextInterpolate1(" ", ctx.title, " ");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("selectedIndex", ctx.index);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.pages);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("disabled", ctx.index == 0);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", ctx.index == ctx.totalPages - 1)("ngIfThen", _r6)("ngIfElse", _r4);
    } }, directives: [ɵngcc2.MatSidenavContainer, ɵngcc2.MatSidenav, ɵngcc3.MatToolbar, ɵngcc4.MatVerticalStepper, ɵngcc5.NgForOf, ɵngcc2.MatSidenavContent, ɵngcc6.MatButton, ɵngcc7.MatIcon, ɵngcc8.MatTabGroup, ɵngcc5.NgIf, ɵngcc4.MatStep, ɵngcc5.NgTemplateOutlet, ɵngcc8.MatTab, ɵngcc8.MatTabContent], styles: [".navigation-buttons[_ngcontent-%COMP%]{margin:1rem;place-content:stretch space-between;-webkit-box-align:stretch;align-items:stretch;-webkit-box-orient:horizontal;-webkit-box-direction:normal;flex-direction:row;box-sizing:border-box;display:-webkit-box;display:flex}"] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(MatPagesComponent, [{
        type: core.Component,
        args: [{
                selector: 'mat-pages',
                template: "<mat-sidenav-container>\n  <mat-sidenav #sidenav [mode]=\"navMode\" opened=\"true\"\n               [fixedInViewport]=\"'false'\" [fixedTopGap]=\"'false'\" [fixedBottomGap]=\"'false'\">\n\n    <mat-toolbar color=\"primary\">\n    </mat-toolbar>\n\n    <mat-vertical-stepper [selectedIndex]=\"index\" (selectionChange)=\"onStepSelectionChange($event)\">\n      <mat-step *ngFor=\"let page of pages;let i = index;\" [label]=\"page.pageOutlook.title\">\n        <ng-container [ngTemplateOutlet]=\"page.pageOutlook.content\"></ng-container>\n      </mat-step>\n    </mat-vertical-stepper>\n  </mat-sidenav>\n\n  <mat-sidenav-content>\n    <mat-toolbar color=\"primary\">\n      <button mat-icon-button (click)=\"sidenav.toggle()\">\n        <mat-icon>menu</mat-icon>\n      </button>\n      {{title}}\n    </mat-toolbar>\n    <div class=\"demo-sidenav-content main-container\">\n\n      <mat-tab-group [selectedIndex]=\"index\" style=\"min-height: 15vh;\" >\n        <button mat-fab color=\"warn\" (click)=\"previous()\">\n          <mat-icon>location</mat-icon>\n        </button>\n        <mat-tab *ngFor=\"let page of pages;let i = index;\">\n          <ng-template matTabContent>\n            <ng-container [ngTemplateOutlet]=\"page.pageContent.content\"></ng-container>\n          </ng-template>\n        </mat-tab>\n      </mat-tab-group>\n\n\n      <div class=\"navigation-buttons\">\n        <button mat-fab color=\"primary\" (click)=\"previous()\" [disabled]=\"index == 0\">\n          <mat-icon>chevron_left</mat-icon>\n        </button>\n        <div *ngIf=\"index == totalPages - 1;then finished else nextAvailable\">\n        </div>\n        <ng-template #nextAvailable>\n          <button mat-fab color=\"primary\" (click)=\"next()\">\n            <mat-icon>chevron_right</mat-icon>\n          </button>\n        </ng-template>\n        <ng-template #finished>\n          <button mat-fab color=\"accent\" (click)=\"done.next()\">\n            <mat-icon>done</mat-icon>\n          </button>\n        </ng-template>\n      </div>\n\n    </div>\n  </mat-sidenav-content>\n</mat-sidenav-container>\n",
                styles: [".navigation-buttons{margin:1rem;place-content:stretch space-between;-webkit-box-align:stretch;align-items:stretch;-webkit-box-orient:horizontal;-webkit-box-direction:normal;flex-direction:row;box-sizing:border-box;display:-webkit-box;display:flex}"]
            }]
    }], function () { return [{ type: ɵngcc0.Renderer2 }]; }, { done: [{
            type: core.Output
        }], pageChanged: [{
            type: core.Output
        }], onResize: [{
            type: core.HostListener,
            args: ['window:resize', ['$event']]
        }], sidenav: [{
            type: core.ViewChild,
            args: ['sidenav']
        }], tabHeader: [{
            type: core.ViewChild,
            args: [tabs.MatTabGroup, { read: core.ElementRef, static: true }]
        }], pages: [{
            type: core.ContentChildren,
            args: [MatPageLoaderComponent]
        }], title: [{
            type: core.Input
        }] }); })();
        return MatPagesComponent;
    }());

    var MatPagesModule = /** @class */ (function () {
        function MatPagesModule() {
        }
        MatPagesModule_1 = MatPagesModule;
        MatPagesModule.forRoot = function () {
            return {
                ngModule: MatPagesModule_1,
                providers: []
            };
        };
        var MatPagesModule_1;
MatPagesModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: MatPagesModule });
MatPagesModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function MatPagesModule_Factory(t) { return new (t || MatPagesModule)(); }, imports: [[
            common.CommonModule,
            toolbar.MatToolbarModule,
            tabs.MatTabsModule,
            card.MatCardModule,
            sidenav.MatSidenavModule,
            icon.MatIconModule,
            stepper.MatStepperModule,
            button.MatButtonModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(MatPagesModule, { declarations: [MatPagesComponent,
        MatPageLoaderComponent,
        MatPageOutlookComponent,
        MatPageContentComponent], imports: [ɵngcc5.CommonModule, ɵngcc3.MatToolbarModule, ɵngcc8.MatTabsModule, ɵngcc1.MatCardModule, ɵngcc2.MatSidenavModule, ɵngcc7.MatIconModule, ɵngcc4.MatStepperModule, ɵngcc6.MatButtonModule], exports: [MatPagesComponent,
        MatPageLoaderComponent,
        MatPageOutlookComponent,
        MatPageContentComponent] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(MatPagesModule, [{
        type: core.NgModule,
        args: [{
                imports: [
                    common.CommonModule,
                    toolbar.MatToolbarModule,
                    tabs.MatTabsModule,
                    card.MatCardModule,
                    sidenav.MatSidenavModule,
                    icon.MatIconModule,
                    stepper.MatStepperModule,
                    button.MatButtonModule
                ],
                exports: [
                    MatPagesComponent,
                    MatPageLoaderComponent,
                    MatPageOutlookComponent,
                    MatPageContentComponent
                ],
                declarations: [
                    MatPagesComponent,
                    MatPageLoaderComponent,
                    MatPageOutlookComponent,
                    MatPageContentComponent
                ]
            }]
    }], function () { return []; }, null); })();
        return MatPagesModule;
    }());

    exports.MatPageContentComponent = MatPageContentComponent;
    exports.MatPageLoaderComponent = MatPageLoaderComponent;
    exports.MatPageOutlookComponent = MatPageOutlookComponent;
    exports.MatPagesComponent = MatPagesComponent;
    exports.MatPagesModule = MatPagesModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
