// @alyle/ui/__ivy_ngcc__/fesm2015/alyle-ui-skeleton.js
import { Directive, Input, NgModule } from '@angular/core';
import { keyframesUniqueId, Dir, toBoolean, StyleRenderer } from '@alyle/ui';

import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@alyle/ui';
const STYLE_PRIORITY = -0.5;
const STYLES = (theme) => {
    const id = keyframesUniqueId.next();
    const { primary, secondary, tertiary } = theme.background;
    const dir = theme.direction === Dir.ltr ? -1 : 1;
    const lum = primary.default.luminance();
    let one = (lum < .5
        ? tertiary
        : secondary);
    let two = (lum < .5
        ? secondary
        : tertiary);
    one = one.darken(1 * (lum < .5 ? -.5 : 0));
    two = two.darken(.25 * (lum < .5 ? -1 : 1));
    return {
        $name: LySkeleton.и,
        $priority: STYLE_PRIORITY,
        $global: (className) => `@keyframes ${id}{${className} 0%{background-position:${-dir * 200}% 50%;}${className} 100%{background-position:${dir * 200}% 50%;}}`,
        root: (className) => `${className}{content:'';background:${`linear-gradient(270deg, ${one}, ${two}, ${two}, ${one})`};background-size:400% 400%;animation:${id} 8s ease-in-out infinite;color:transparent;cursor:progress;user-select:none;}`
    };
};
class LySkeleton {
    constructor(sRenderer) {
        this.sRenderer = sRenderer;
        /** @docs-private */
        this.classes = this.sRenderer.renderSheet(STYLES);
    }
    get skeleton() {
        return this._skeleton;
    }
    set skeleton(val) {
        const newVal = toBoolean(val);
        this._skeleton = newVal;
        this.sRenderer.toggleClass(this.classes.root, newVal);
    }
}
LySkeleton.ɵfac = function LySkeleton_Factory(t) { return new (t || LySkeleton)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.StyleRenderer)); };
LySkeleton.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: LySkeleton, selectors: [["", "lySkeleton", ""]], inputs: { skeleton: ["lySkeleton", "skeleton"] }, exportAs: ["lySkeleton"], features: [ɵngcc0.ɵɵProvidersFeature([
            StyleRenderer
        ])] });
/** @docs-private */
LySkeleton.и = 'LySkeleton';
LySkeleton.ctorParameters = () => [
    { type: StyleRenderer }
];
LySkeleton.propDecorators = {
    skeleton: [{ type: Input, args: ['lySkeleton',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(LySkeleton, [{
        type: Directive,
        args: [{
                selector: '[lySkeleton]',
                providers: [
                    StyleRenderer
                ],
                exportAs: 'lySkeleton'
            }]
    }], function () { return [{ type: ɵngcc1.StyleRenderer }]; }, { skeleton: [{
            type: Input,
            args: ['lySkeleton']
        }] }); })();

class LySkeletonModule {
}
LySkeletonModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: LySkeletonModule });
LySkeletonModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function LySkeletonModule_Factory(t) { return new (t || LySkeletonModule)(); } });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(LySkeletonModule, { declarations: [LySkeleton], exports: [LySkeleton] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(LySkeletonModule, [{
        type: NgModule,
        args: [{
                declarations: [LySkeleton],
                exports: [LySkeleton]
            }]
    }], null, null); })();

/**
 * Generated bundle index. Do not edit.
 */

export { LySkeleton, LySkeletonModule, STYLES };
