// @alyle/ui/esm2015/field/display-with.js
import { Directive, Inject } from '@angular/core';
import { StyleRenderer } from '@alyle/ui';
import { LY_FIELD_STYLES_TOKEN } from './field-styles-token';
/**
 * Only show when input is not focused, and hides input`.
 * You can use it to display formatted values.
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@alyle/ui';
export class LyDisplayWith {
    constructor(sRenderer, styles) {
        this.sRenderer = sRenderer;
        sRenderer.renderSheet(styles, 'displayWith');
    }
}
LyDisplayWith.ɵfac = function LyDisplayWith_Factory(t) { return new (t || LyDisplayWith)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.StyleRenderer), ɵngcc0.ɵɵdirectiveInject(LY_FIELD_STYLES_TOKEN)); };
LyDisplayWith.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: LyDisplayWith, selectors: [["ly-display-with"]], features: [ɵngcc0.ɵɵProvidersFeature([
            StyleRenderer
        ])] });
LyDisplayWith.ctorParameters = () => [
    { type: StyleRenderer },
    { type: undefined, decorators: [{ type: Inject, args: [LY_FIELD_STYLES_TOKEN,] }] }
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(LyDisplayWith, [{
        type: Directive,
        args: [{
                selector: 'ly-display-with',
                providers: [
                    StyleRenderer
                ]
            }]
    }], function () { return [{ type: ɵngcc1.StyleRenderer }, { type: undefined, decorators: [{
                type: Inject,
                args: [LY_FIELD_STYLES_TOKEN]
            }] }]; }, null); })();
