// @alyle/ui/esm2015/field/field.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { LyCommonModule } from '@alyle/ui';
import { ObserversModule } from '@angular/cdk/observers';
import { LyField, LyNativeControl, STYLES } from './field';
import { LyPlaceholder } from './placeholder';
import { LyLabel } from './label';
import { LyPrefix } from './prefix';
import { LySuffix } from './suffix';
import { LyHint } from './hint';
import { LyError } from './error';
import { LY_FIELD_STYLES_TOKEN } from './field-styles-token';
import { LyDisplayWith } from './display-with';
import * as ɵngcc0 from '@angular/core';
const ɵ0 = STYLES;
export class LyFieldModule {
}
LyFieldModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: LyFieldModule });
LyFieldModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function LyFieldModule_Factory(t) { return new (t || LyFieldModule)(); }, providers: [
        {
            provide: LY_FIELD_STYLES_TOKEN,
            useValue: ɵ0
        }
    ], imports: [[
            CommonModule,
            LyCommonModule,
            ObserversModule
        ],
        LyCommonModule] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(LyFieldModule, { declarations: function () { return [LyField,
        LyPlaceholder,
        LyLabel,
        LyNativeControl,
        LyPrefix,
        LySuffix,
        LyHint,
        LyError,
        LyDisplayWith]; }, imports: function () { return [CommonModule,
        LyCommonModule,
        ObserversModule]; }, exports: function () { return [LyField,
        LyPlaceholder,
        LyLabel,
        LyNativeControl,
        LyPrefix,
        LySuffix,
        LyHint,
        LyError,
        LyDisplayWith,
        LyCommonModule]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(LyFieldModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    LyCommonModule,
                    ObserversModule
                ],
                exports: [
                    LyField,
                    LyPlaceholder,
                    LyLabel,
                    LyNativeControl,
                    LyPrefix,
                    LySuffix,
                    LyHint,
                    LyError,
                    LyDisplayWith,
                    LyCommonModule
                ],
                providers: [
                    {
                        provide: LY_FIELD_STYLES_TOKEN,
                        useValue: ɵ0
                    }
                ],
                declarations: [
                    LyField,
                    LyPlaceholder,
                    LyLabel,
                    LyNativeControl,
                    LyPrefix,
                    LySuffix,
                    LyHint,
                    LyError,
                    LyDisplayWith
                ]
            }]
    }], null, null); })();
export { ɵ0 };
