// @angular-material-extensions/pages/esm5/lib/mat-page-outlook/mat-page-outlook.component.js
import { __decorate, __metadata } from "tslib";
import { ChangeDetectionStrategy, Component, Input, TemplateRef, ViewChild, ViewEncapsulation } from '@angular/core';
import * as ɵngcc0 from '@angular/core';

function MatPageOutlookComponent_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵprojection(0);
} }
var _c0 = ["*"];
var MatPageOutlookComponent = /** @class */ (function () {
    function MatPageOutlookComponent() {
    }
    MatPageOutlookComponent.prototype.ngOnInit = function () {
    };
    __decorate([
        ViewChild(TemplateRef),
        __metadata("design:type", TemplateRef)
    ], MatPageOutlookComponent.prototype, "content", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], MatPageOutlookComponent.prototype, "title", void 0);
    MatPageOutlookComponent = __decorate([ __metadata("design:paramtypes", [])
    ], MatPageOutlookComponent);
MatPageOutlookComponent.ɵfac = function MatPageOutlookComponent_Factory(t) { return new (t || MatPageOutlookComponent)(); };
MatPageOutlookComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: MatPageOutlookComponent, selectors: [["mat-page-outlook"]], viewQuery: function MatPageOutlookComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(TemplateRef, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.content = _t.first);
    } }, inputs: { title: "title" }, ngContentSelectors: _c0, decls: 1, vars: 0, template: function MatPageOutlookComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵtemplate(0, MatPageOutlookComponent_ng_template_0_Template, 1, 0, "ng-template");
    } }, styles: [""], encapsulation: 2, changeDetection: 0 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(MatPageOutlookComponent, [{
        type: Component,
        args: [{
                selector: 'mat-page-outlook',
                template: "<ng-template>\n  <ng-content></ng-content>\n</ng-template>\n",
                encapsulation: ViewEncapsulation.None,
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [""]
            }]
    }], function () { return []; }, { content: [{
            type: ViewChild,
            args: [TemplateRef]
        }], title: [{
            type: Input
        }] }); })();
    return MatPageOutlookComponent;
}());
export { MatPageOutlookComponent };
