// @alyle/ui/esm2015/select/select.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { LyCommonModule, LyOverlayModule } from '@alyle/ui';
import { LyCheckboxModule } from '@alyle/ui/checkbox';
import { LySelect, LyOption, LySelectTrigger } from './select';
import * as ɵngcc0 from '@angular/core';
export class LySelectModule {
}
LySelectModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: LySelectModule });
LySelectModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function LySelectModule_Factory(t) { return new (t || LySelectModule)(); }, imports: [[
            CommonModule,
            LyCommonModule,
            LyCheckboxModule,
            LyOverlayModule
        ],
        LyCommonModule] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(LySelectModule, { declarations: function () { return [LySelect, LyOption, LySelectTrigger]; }, imports: function () { return [CommonModule,
        LyCommonModule,
        LyCheckboxModule,
        LyOverlayModule]; }, exports: function () { return [LySelect, LyOption, LySelectTrigger, LyCommonModule]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(LySelectModule, [{
        type: NgModule,
        args: [{
                declarations: [LySelect, LyOption, LySelectTrigger],
                imports: [
                    CommonModule,
                    LyCommonModule,
                    LyCheckboxModule,
                    LyOverlayModule
                ],
                exports: [LySelect, LyOption, LySelectTrigger, LyCommonModule]
            }]
    }], null, null); })();
