// @angular-material-extensions/pages/esm2015/lib/mat-pages.module.js
import * as ɵngcc0 from '@angular/core';
var MatPagesModule_1;
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { MatButtonModule } from '@angular/material/button';
import { MatCardModule } from '@angular/material/card';
import { MatIconModule } from '@angular/material/icon';
import { MatSidenavModule } from '@angular/material/sidenav';
import { MatStepperModule } from '@angular/material/stepper';
import { MatTabsModule } from '@angular/material/tabs';
import { MatToolbarModule } from '@angular/material/toolbar';
import { MatPagesComponent } from './mat-pages.component';
import { MatPageLoaderComponent } from './mat-page-loader/mat-page-loader.component';
import { MatPageContentComponent } from './mat-page-content/mat-page-content.component';
import { MatPageOutlookComponent } from './mat-page-outlook/mat-page-outlook.component';
let MatPagesModule = MatPagesModule_1 = class MatPagesModule {
    static forRoot() {
        return {
            ngModule: MatPagesModule_1,
            providers: []
        };
    }
};
MatPagesModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: MatPagesModule });
MatPagesModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function MatPagesModule_Factory(t) { return new (t || MatPagesModule)(); }, imports: [[
            CommonModule,
            MatToolbarModule,
            MatTabsModule,
            MatCardModule,
            MatSidenavModule,
            MatIconModule,
            MatStepperModule,
            MatButtonModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(MatPagesModule, { declarations: function () { return [MatPagesComponent,
        MatPageLoaderComponent,
        MatPageOutlookComponent,
        MatPageContentComponent]; }, imports: function () { return [CommonModule,
        MatToolbarModule,
        MatTabsModule,
        MatCardModule,
        MatSidenavModule,
        MatIconModule,
        MatStepperModule,
        MatButtonModule]; }, exports: function () { return [MatPagesComponent,
        MatPageLoaderComponent,
        MatPageOutlookComponent,
        MatPageContentComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(MatPagesModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    MatToolbarModule,
                    MatTabsModule,
                    MatCardModule,
                    MatSidenavModule,
                    MatIconModule,
                    MatStepperModule,
                    MatButtonModule
                ],
                exports: [
                    MatPagesComponent,
                    MatPageLoaderComponent,
                    MatPageOutlookComponent,
                    MatPageContentComponent
                ],
                declarations: [
                    MatPagesComponent,
                    MatPageLoaderComponent,
                    MatPageOutlookComponent,
                    MatPageContentComponent
                ]
            }]
    }], null, null); })();
export { MatPagesModule };
