// @alyle/ui/esm2015/image-cropper/image-cropper.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { LyImageCropper } from './image-cropper';
import { LyCropperArea } from './image-cropper-area';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
export class LyImageCropperModule {
}
LyImageCropperModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: LyImageCropperModule });
LyImageCropperModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function LyImageCropperModule_Factory(t) { return new (t || LyImageCropperModule)(); }, imports: [[CommonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(LyImageCropperModule, { declarations: function () { return [LyImageCropper, LyCropperArea]; }, imports: function () { return [CommonModule]; }, exports: function () { return [LyImageCropper]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(LyImageCropperModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                exports: [LyImageCropper],
                declarations: [LyImageCropper, LyCropperArea]
            }]
    }], null, null); })();
ɵngcc0.ɵɵsetComponentScope(LyImageCropper, [ɵngcc1.NgClass, ɵngcc1.NgComponentOutlet, ɵngcc1.NgForOf, ɵngcc1.NgIf, ɵngcc1.NgTemplateOutlet, ɵngcc1.NgStyle, ɵngcc1.NgSwitch, ɵngcc1.NgSwitchCase, ɵngcc1.NgSwitchDefault, ɵngcc1.NgPlural, ɵngcc1.NgPluralCase, LyImageCropper, LyCropperArea], [ɵngcc1.AsyncPipe, ɵngcc1.UpperCasePipe, ɵngcc1.LowerCasePipe, ɵngcc1.JsonPipe, ɵngcc1.SlicePipe, ɵngcc1.DecimalPipe, ɵngcc1.PercentPipe, ɵngcc1.TitleCasePipe, ɵngcc1.CurrencyPipe, ɵngcc1.DatePipe, ɵngcc1.I18nPluralPipe, ɵngcc1.I18nSelectPipe, ɵngcc1.KeyValuePipe]);
