// ngx-intl-tel-input/esm5/lib/directives/native-element-injector.directive.js
import { __decorate, __metadata } from "tslib";
import { Directive, ElementRef, OnInit } from '@angular/core';
import { NgControl } from '@angular/forms';
/*
"Property 'nativeElement' does not exist on type 'FormControl'".
'NativeElementInjectorDirective' injects nativeElement to each control,
so we can access it from inside validator for example.
More about this approach and reasons for this:
https://github.com/angular/angular/issues/18025
https://stackoverflow.com/a/54075119/1617590
*/
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/forms';
var NativeElementInjectorDirective = /** @class */ (function () {
    function NativeElementInjectorDirective(controlDir, host) {
        this.controlDir = controlDir;
        this.host = host;
    }
    NativeElementInjectorDirective.prototype.ngOnInit = function () {
        if (this.controlDir.control) {
            this.controlDir.control['nativeElement'] = this.host.nativeElement;
        }
    };
    NativeElementInjectorDirective.ctorParameters = function () { return [
        { type: NgControl },
        { type: ElementRef }
    ]; };
    NativeElementInjectorDirective = __decorate([ __metadata("design:paramtypes", [NgControl,
            ElementRef])
    ], NativeElementInjectorDirective);
NativeElementInjectorDirective.ɵfac = function NativeElementInjectorDirective_Factory(t) { return new (t || NativeElementInjectorDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.NgControl), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
NativeElementInjectorDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: NativeElementInjectorDirective, selectors: [["", "ngModel", ""], ["", "formControl", ""], ["", "formControlName", ""]] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NativeElementInjectorDirective, [{
        type: Directive,
        args: [{
                // tslint:disable-next-line: directive-selector
                selector: '[ngModel], [formControl], [formControlName]'
            }]
    }], function () { return [{ type: ɵngcc1.NgControl }, { type: ɵngcc0.ElementRef }]; }, null); })();
    return NativeElementInjectorDirective;
}());
export { NativeElementInjectorDirective };
