// @material-extended/mde/esm2015/lib/popover/popover-target.js
import { __decorate } from "tslib";
import { Directive, ElementRef } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
let MdePopoverTarget = class MdePopoverTarget {
    constructor(_elementRef) {
        this._elementRef = _elementRef;
    }
};
MdePopoverTarget.ɵfac = function MdePopoverTarget_Factory(t) { return new (t || MdePopoverTarget)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
MdePopoverTarget.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: MdePopoverTarget, selectors: [["mde-popover-target"], ["", "mdePopoverTarget", ""]], exportAs: ["mdePopoverTarget"] });
MdePopoverTarget.ctorParameters = () => [
    { type: ElementRef }
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(MdePopoverTarget, [{
        type: Directive,
        args: [{
                selector: 'mde-popover-target, [mdePopoverTarget]',
                exportAs: 'mdePopoverTarget'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }]; }, null); })();
export { MdePopoverTarget };
