// @alyle/ui/esm2015/src/theme/common.module.js
import { NgModule } from '@angular/core';
import { LyPaper } from './paper';
import { LyWithClass } from './with-class.directive';
import { LyStyle, LyStyleDeprecated } from './style.directive';
import * as ɵngcc0 from '@angular/core';
export class LyCommonModule {
}
LyCommonModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: LyCommonModule });
LyCommonModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function LyCommonModule_Factory(t) { return new (t || LyCommonModule)(); } });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(LyCommonModule, { declarations: function () { return [LyStyle, LyWithClass, LyPaper]; }, exports: function () { return [LyStyle, LyWithClass, LyPaper]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(LyCommonModule, [{
        type: NgModule,
        args: [{
                declarations: [LyStyle, LyWithClass, LyPaper],
                exports: [LyStyle, LyWithClass, LyPaper]
            }]
    }], null, null); })();
/**
 * @deprecated
 */
export class LyStyleDeprecatedModule {
}
LyStyleDeprecatedModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: LyStyleDeprecatedModule });
LyStyleDeprecatedModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function LyStyleDeprecatedModule_Factory(t) { return new (t || LyStyleDeprecatedModule)(); } });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(LyStyleDeprecatedModule, { declarations: function () { return [LyStyleDeprecated]; }, exports: function () { return [LyStyleDeprecated]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(LyStyleDeprecatedModule, [{
        type: NgModule,
        args: [{
                exports: [LyStyleDeprecated],
                declarations: [LyStyleDeprecated]
            }]
    }], null, null); })();
