// @alyle/ui/esm2015/src/theme/with-class.directive.js
import { Directive, ElementRef, Input } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
export class LyWithClass {
    constructor(el) {
        this.el = el;
    }
    set withClass(val) {
        if (!val) {
            throw new Error(`'${val}' is not valid className`);
        }
        this.el.nativeElement.classList.add(val);
    }
}
LyWithClass.ɵfac = function LyWithClass_Factory(t) { return new (t || LyWithClass)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
LyWithClass.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: LyWithClass, selectors: [["", "withClass", ""]], inputs: { withClass: "withClass" } });
LyWithClass.ctorParameters = () => [
    { type: ElementRef }
];
LyWithClass.propDecorators = {
    withClass: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(LyWithClass, [{
        type: Directive,
        args: [{
                selector: '[withClass]'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }]; }, { withClass: [{
            type: Input
        }] }); })();
