// @alyle/ui/esm2015/toolbar/toolbar.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { LyCommonModule } from '@alyle/ui';
import { LyToolbar } from './toolbar';
import * as ɵngcc0 from '@angular/core';
export class LyToolbarModule {
}
LyToolbarModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: LyToolbarModule });
LyToolbarModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function LyToolbarModule_Factory(t) { return new (t || LyToolbarModule)(); }, imports: [[CommonModule, LyCommonModule],
        LyCommonModule] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(LyToolbarModule, { declarations: function () { return [LyToolbar]; }, imports: function () { return [CommonModule, LyCommonModule]; }, exports: function () { return [LyToolbar, LyCommonModule]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(LyToolbarModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, LyCommonModule],
                exports: [LyToolbar, LyCommonModule],
                declarations: [LyToolbar]
            }]
    }], null, null); })();
