// @alyle/ui/bundles/alyle-ui-carousel.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('@alyle/ui'),require('@angular/cdk/platform'),require('@angular/common'),exports, require('@angular/core'), require('@alyle/ui'), require('rxjs'), require('rxjs/operators'), require('@angular/cdk/platform'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('@alyle/ui/carousel', ['@angular/core','@alyle/ui','@angular/cdk/platform','@angular/common','exports', '@angular/core', '@alyle/ui', 'rxjs', 'rxjs/operators', '@angular/cdk/platform', '@angular/common'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.ng.core,global.alyle.ui,global.ng.cdk.platform,global.ng.common,(global.ly = global.ly || {}, global.ly.carousel = {}), global.ng.core, global.ly.core, global.rxjs, global.rxjs.operators, global.ng.cdk.platform, global.ng.common));
}(this, (function (ɵngcc0,ɵngcc1,ɵngcc2,ɵngcc3,exports, core, ui, rxjs, operators, platform, common) { 
var _c0 = ["slideContainer"];
var _c1 = ["_slide"];
var _c2 = ["_progressBar"];
var _c3 = function (a0, a1) { return [a0, a1]; };
function LyCarousel_ol_5_li_1_Template(rf, ctx) { if (rf & 1) {
    var _r10 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "li", 7);
    ɵngcc0.ɵɵlistener("click", function LyCarousel_ol_5_li_1_Template_li_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r10); var i_r8 = ctx.index; var ctx_r9 = ɵngcc0.ɵɵnextContext(2); return ctx_r9._select(i_r8); });
    ɵngcc0.ɵɵelement(1, "span");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var i_r8 = ctx.index;
    var ctx_r6 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction2(3, _c3, ctx_r6.classes.indicator, ctx_r6.selectedIndex == i_r8 ? ctx_r6.classes.indicatorActive : ""));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵclassMap(ctx_r6.classes.indicatorIcon);
} }
function LyCarousel_ol_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "ol", 2);
    ɵngcc0.ɵɵtemplate(1, LyCarousel_ol_5_li_1_Template, 2, 6, "li", 6);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("className", ctx_r2.classes.indicators);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r2.lyItems);
} }
var _c4 = function (a0) { return [a0, "left"]; };
function LyCarousel_div_6_Template(rf, ctx) { if (rf & 1) {
    var _r12 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 7);
    ɵngcc0.ɵɵlistener("click", function LyCarousel_div_6_Template_div_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r12); var ctx_r11 = ɵngcc0.ɵɵnextContext(); return ctx_r11.prev(); });
    ɵngcc0.ɵɵnamespaceSVG();
    ɵngcc0.ɵɵelementStart(1, "svg", 8);
    ɵngcc0.ɵɵelement(2, "path", 9);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(1, _c4, ctx_r3.classes.actions));
} }
var _c5 = function (a0) { return [a0, "right"]; };
function LyCarousel_div_7_Template(rf, ctx) { if (rf & 1) {
    var _r14 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 7);
    ɵngcc0.ɵɵlistener("click", function LyCarousel_div_7_Template_div_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r14); var ctx_r13 = ɵngcc0.ɵɵnextContext(); return ctx_r13.next(); });
    ɵngcc0.ɵɵnamespaceSVG();
    ɵngcc0.ɵɵelementStart(1, "svg", 8);
    ɵngcc0.ɵɵelement(2, "path", 9);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r4 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(1, _c5, ctx_r4.classes.actions));
} }
function LyCarousel_div_8_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 2);
    ɵngcc0.ɵɵelement(1, "div", 2, 10);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r5 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("className", ctx_r5.classes.barContainer);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵstyleProp("animation-duration", ctx_r5.interval + "ms");
    ɵngcc0.ɵɵproperty("className", ctx_r5.classes.bar);
} }
var _c6 = ["*"];
'use strict';

    /** Default interval in ms */
    var DEFAULT_INTERVAL = 7000;
    var DEFAULT_AUTOPLAY = true;
    var DEFAULT_HAS_PROGRESS_BAR = false;
    var STYLE_PRIORITY = -2;
    var STYLES = function (theme, ref) {
        var dir = theme.getDirection(ui.DirAlias.before);
        var right = dir === 'right' ? 0 : 180;
        var left = dir === 'left' ? 0 : 180;
        var carousel = ref.selectorsOf(STYLES);
        var barAnimation = ui.keyframesUniqueId.next();
        var after = theme.after, before = theme.before;
        return {
            $priority: STYLE_PRIORITY,
            $global: function (className) { return "@keyframes " + barAnimation + "{" + className + " 0%{transform:translateX(0%);}" + className + " 100%{transform:translateX(" + (dir === 'left' ? '-' : '') + "100%);}}"; },
            root: function () { return function (className) { return className + "{display:block;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;position:relative;}" + ui.st2c(((theme.carousel
                && theme.carousel.root
                && (theme.carousel.root instanceof ui.StyleCollection
                    ? theme.carousel.root.setTransformer(function (fn) { return fn(carousel); })
                    : theme.carousel.root(carousel)))), "" + className) + className + " " + carousel.actions + ".right{" + after + ":0;transform:rotate(" + right + "deg);}" + className + " " + carousel.actions + ".left{" + before + ":0;transform:rotate(" + left + "deg);}" + className + " svg{display:block;fill:currentColor;}"; }; },
            actions: function (className) { return className + "{position:absolute;top:0;bottom:0;margin:auto .25em;height:1em;width:1em;font-size:36px;cursor:pointer;background:" + theme.background.primary.default.alpha(.25) + ";color:" + theme.text.primary + ";will-change:transform;}"; },
            slideContainer: function (className) { return className + "{overflow:hidden;display:block;width:100%;height:100%;position:relative;touch-action:pan-y !important;}"; },
            slide: function (className) { return className + "{display:flex;width:100%;height:100%;will-change:transform;}" + className + " > ly-carousel-item{width:100%;flex-shrink:0;position:relative;background-size:cover;background-position:center;background-repeat:no-repeat;}"; },
            slideContent: function (className) { return className + "{display:flex;}"; },
            slideAnim: function (className) { return className + " > div{transition:transform 750ms cubic-bezier(.1, 1, 0.5, 1);}"; },
            slideNoEvent: function (className) { return className + ">div{touch-action:initial !important;-webkit-user-drag:initial !important;}"; },
            indicators: function () { return function (className) { return className + "{position:absolute;bottom:0;left:0;right:0;margin:0;box-sizing:border-box;display:flex;align-items:center;justify-content:center;height:48px;}"; }; },
            indicator: function () { return function (className) { return className + "{display:inline-block;border-radius:50%;cursor:pointer;position:relative;padding:.5em;outline:none;}"; }; },
            indicatorIcon: function () { return function (className) { return className + "{transition:300ms cubic-bezier(0.65, 0.05, 0.36, 1);width:1em;height:1em;transform:scale(.5);border-radius:50%;will-change:transform;display:block;opacity:.65;box-shadow:" + ui.shadowBuilder(8, theme.text.default) + ";background:" + theme.background.primary.default + ";}"; }; },
            indicatorActive: function () { return function (className) { return className + " " + carousel.indicatorIcon + "{transform:scale(1);opacity:1;}"; }; },
            barContainer: function (className) { return className + "{background:" + theme.background.primary.default.alpha(.25) + ";height:4px;position:absolute;bottom:0;width:100%;}"; },
            bar: function (className) { return className + "{height:4px;position:absolute;bottom:0;width:100%;animation-name:" + barAnimation + ";animation-timing-function:linear;animation-iteration-count:infinite;background:" + theme.text.primary + ";}"; }
        };
    };
    (function (CarouselMode) {
        /** full */
        CarouselMode[CarouselMode["default"] = 0] = "default";
        CarouselMode[CarouselMode["inline"] = 1] = "inline";
    })(exports.CarouselMode || (exports.CarouselMode = {}));
    var LyCarousel = /** @class */ (function () {
        function LyCarousel(_el, _cd, _theme, _renderer, sRenderer, _platform) {
            this._el = _el;
            this._cd = _cd;
            this._theme = _theme;
            this._renderer = _renderer;
            this.sRenderer = sRenderer;
            this._platform = _platform;
            /** @docs-private */
            this.classes = this.sRenderer.renderSheet(STYLES, true);
            this._intervalFn = null;
            /** @docs-private */
            this.mode = exports.CarouselMode.default;
            this.selectedIndex = 0;
            this._interval = DEFAULT_INTERVAL;
            /** Emits whenever the component is destroyed. */
            this._destroy = new rxjs.Subject();
            this._hasNavigationArrows = true;
            this._hasNavigationIndicators = true;
        }
        Object.defineProperty(LyCarousel.prototype, "_isIntervalFn", {
            /** @internal */
            get: function () {
                return !!this._intervalFn;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(LyCarousel.prototype, "pauseOnHover", {
            /**
             * It will pause the slide change when the mouse cursor passes
             * through the carousel.
             */
            get: function () {
                return this._pauseOnHover;
            },
            set: function (val) {
                var newVal = ui.toBoolean(val);
                this._pauseOnHover = newVal;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(LyCarousel.prototype, "touch", {
            get: function () {
                return this._touch;
            },
            set: function (val) {
                var newVal = ui.toBoolean(val);
                this._touch = newVal;
                if (newVal) {
                    this._renderer.removeClass(this._el.nativeElement, this.classes.slideNoEvent);
                }
                else {
                    this._renderer.addClass(this._el.nativeElement, this.classes.slideNoEvent);
                }
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(LyCarousel.prototype, "autoplay", {
            get: function () {
                return this._autoplay;
            },
            set: function (val) {
                var newVal = ui.toBoolean(val);
                this._autoplay = newVal;
                if (newVal) {
                    this._resetInterval();
                }
                else {
                    this.stop();
                }
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(LyCarousel.prototype, "hasProgressBar", {
            get: function () {
                return this._hasProgressBar;
            },
            set: function (val) {
                var newVal = ui.toBoolean(val);
                this._hasProgressBar = newVal;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(LyCarousel.prototype, "interval", {
            get: function () {
                return this._interval;
            },
            set: function (val) {
                this._interval = val;
                this._resetInterval();
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(LyCarousel.prototype, "hasNavigationArrows", {
            get: function () {
                return this._hasNavigationArrows;
            },
            set: function (val) {
                this._hasNavigationArrows = ui.toBoolean(val);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(LyCarousel.prototype, "hasNavigationIndicators", {
            get: function () {
                return this._hasNavigationIndicators;
            },
            set: function (val) {
                this._hasNavigationIndicators = ui.toBoolean(val);
            },
            enumerable: false,
            configurable: true
        });
        LyCarousel.prototype.ngOnInit = function () {
            if (!this.touch) {
                this.touch = false;
            }
            if (this.autoplay == null) {
                this.autoplay = DEFAULT_AUTOPLAY;
            }
            if (this.hasProgressBar == null) {
                this.hasProgressBar = DEFAULT_HAS_PROGRESS_BAR;
            }
        };
        LyCarousel.prototype.ngAfterViewInit = function () {
            var _this = this;
            this._renderer.addClass(this.slideContainer.nativeElement, this.classes.slideContainer);
            if (this._platform.isBrowser) {
                this._renderer.addClass(this.slideContainer.nativeElement, this.classes.slideAnim);
            }
            this.lyItems.changes.pipe(operators.takeUntil(this._destroy)).subscribe(function () { return _this._markForCheck(); });
        };
        LyCarousel.prototype.ngOnDestroy = function () {
            this._destroy.next();
            this._destroy.complete();
            if (this._platform.isBrowser) {
                this.stop();
            }
        };
        LyCarousel.prototype._onMouseEnter = function () {
            if (this.pauseOnHover) {
                this.stop();
            }
        };
        LyCarousel.prototype._onMouseLeave = function () {
            if (this.pauseOnHover) {
                this._resetInterval();
            }
        };
        /** @docs-private */
        LyCarousel.prototype._onDragStart = function () {
            var _this = this;
            this.stop();
            this._renderer.removeClass(this.slideContainer.nativeElement, this.classes.slideAnim);
            this._selectedElement = this.lyItems.find(function (_item, index) { return index === _this.selectedIndex; })._nativeElement;
        };
        LyCarousel.prototype._onDrag = function (e) {
            var rect = this._selectedElement.getBoundingClientRect();
            if (Math.abs(e.deltaX) < rect.width) {
                this._onPan(e.deltaX);
            }
            else {
                this._onPan(rect.width * Math.sign(e.deltaX));
            }
        };
        LyCarousel.prototype._onDragEnd = function (e) {
            var rect = this._selectedElement.getBoundingClientRect();
            var dir = this._theme.variables.getDirection(ui.DirAlias.before);
            this._renderer.addClass(this.slideContainer.nativeElement, this.classes.slideAnim);
            this._select(this.selectedIndex);
            if (Math.abs(e.deltaX) > rect.width / 2) {
                if (0 > e.deltaX) {
                    this.next();
                }
                else if (0 < e.deltaX) {
                    this.prev();
                }
            }
            else if (e.additionalEvent) {
                var eventName = e.additionalEvent;
                if (Math.abs(e.velocity) >= 0.25) {
                    if (eventName === 'slideleft') {
                        if (dir === 'left') {
                            this.next();
                        }
                        else {
                            this.prev();
                        }
                    }
                    else if (eventName === 'slideright') {
                        if (dir === 'right') {
                            this.next();
                        }
                        else {
                            this.prev();
                        }
                    }
                }
            }
            this._renderer.removeStyle(this._slide.nativeElement, 'transform');
        };
        LyCarousel.prototype._onDragCancel = function () {
            this._renderer.addClass(this.slideContainer.nativeElement, this.classes.slideAnim);
            this._select(this.selectedIndex);
            this._resetInterval();
        };
        LyCarousel.prototype._select = function (val, notResetInterval) {
            this.selectedIndex = val;
            if (this.mode === exports.CarouselMode.default) {
                this._slideClass = this._theme.addStyle("lyCarousel.select:" + val.toString(32), function (theme) {
                    var sign = theme.getDirection(ui.DirAlias.before) === 'left' ? -1 : 1;
                    return {
                        transform: "translateX(" + 100 * val * sign + "%)"
                    };
                }, this._slide.nativeElement, this._slideClass, STYLE_PRIORITY);
            }
            if (!notResetInterval) {
                if (this.autoplay && !this.pauseOnHover) {
                    this._resetInterval();
                }
            }
        };
        LyCarousel.prototype.prev = function () {
            var len = this.lyItems.length - 1;
            var prev = this.selectedIndex - 1;
            this._select(prev < 0 ? len : prev);
        };
        LyCarousel.prototype.next = function (notResetInterval) {
            var len = this.lyItems.length - 1;
            var next = this.selectedIndex + 1;
            this._select(next > len ? 0 : next, notResetInterval);
        };
        LyCarousel.prototype.stop = function () {
            if (this._intervalFn !== null) {
                clearInterval(this._intervalFn);
                this._intervalFn = null;
            }
        };
        LyCarousel.prototype._resetInterval = function () {
            var _this = this;
            if (this._platform.isBrowser) {
                this.stop();
                this._restartProgressBarAnimation();
                this._markForCheck();
                this._intervalFn = setInterval(function () {
                    _this.next(true);
                    _this._restartProgressBarAnimation();
                    _this._markForCheck();
                }, this.interval);
            }
        };
        LyCarousel.prototype._restartProgressBarAnimation = function () {
            if (this.hasProgressBar && this._progressBar) {
                var el = this._progressBar.nativeElement;
                // Hack for restart animation
                el.style.animationName = 'øfakeName';
                window.getComputedStyle(el).getPropertyValue('opacity');
                el.style.animationName = '';
            }
        };
        LyCarousel.prototype._onPan = function (x) {
            var sign = this._theme.variables.getDirection(ui.DirAlias.before) === 'left' ? -1 : 1;
            this._renderer.setStyle(this._slide.nativeElement, 'transform', "translateX(calc(" + sign * 100 * this.selectedIndex + "% + " + x + "px))");
        };
        LyCarousel.prototype._markForCheck = function () {
            this._cd.markForCheck();
        };
LyCarousel.ɵfac = function LyCarousel_Factory(t) { return new (t || LyCarousel)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ChangeDetectorRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.LyTheme2), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Renderer2), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.StyleRenderer), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.Platform)); };
LyCarousel.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: LyCarousel, selectors: [["ly-carousel"]], contentQueries: function LyCarousel_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, LyCarouselItem, false);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.lyItems = _t);
    } }, viewQuery: function LyCarousel_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c0, true);
        ɵngcc0.ɵɵviewQuery(_c1, true);
        ɵngcc0.ɵɵviewQuery(_c2, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.slideContainer = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx._slide = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx._progressBar = _t.first);
    } }, hostBindings: function LyCarousel_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("mouseenter", function LyCarousel_mouseenter_HostBindingHandler() { return ctx._onMouseEnter(); })("mouseleave", function LyCarousel_mouseleave_HostBindingHandler() { return ctx._onMouseLeave(); });
    } }, inputs: { mode: "mode", selectedIndex: "selectedIndex", pauseOnHover: "pauseOnHover", touch: "touch", autoplay: "autoplay", hasProgressBar: "hasProgressBar", interval: "interval", hasNavigationArrows: "hasNavigationArrows", hasNavigationIndicators: "hasNavigationIndicators" }, features: [ɵngcc0.ɵɵProvidersFeature([
            ui.StyleRenderer
        ])], ngContentSelectors: _c6, decls: 9, vars: 5, consts: [[3, "slidestart", "slideleft", "slideright", "slidecancel", "slideend"], ["slideContainer", ""], [3, "className"], ["_slide", ""], [3, "className", 4, "ngIf"], [3, "ngClass", "click", 4, "ngIf"], [3, "ngClass", "click", 4, "ngFor", "ngForOf"], [3, "ngClass", "click"], ["viewBox", "0 0 24 24"], ["d", "M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"], ["_progressBar", ""]], template: function LyCarousel_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵelementStart(0, "div", 0, 1);
        ɵngcc0.ɵɵlistener("slidestart", function LyCarousel_Template_div_slidestart_0_listener() { return ctx.touch && ctx._onDragStart(); })("slideleft", function LyCarousel_Template_div_slideleft_0_listener($event) { return ctx.touch && ctx._onDrag($event); })("slideright", function LyCarousel_Template_div_slideright_0_listener($event) { return ctx.touch && ctx._onDrag($event); })("slidecancel", function LyCarousel_Template_div_slidecancel_0_listener() { return ctx.touch && ctx._onDragCancel(); })("slideend", function LyCarousel_Template_div_slideend_0_listener($event) { return ctx.touch && ctx._onDragEnd($event); });
        ɵngcc0.ɵɵelementStart(2, "div", 2, 3);
        ɵngcc0.ɵɵprojection(4);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(5, LyCarousel_ol_5_Template, 2, 2, "ol", 4);
        ɵngcc0.ɵɵtemplate(6, LyCarousel_div_6_Template, 3, 3, "div", 5);
        ɵngcc0.ɵɵtemplate(7, LyCarousel_div_7_Template, 3, 3, "div", 5);
        ɵngcc0.ɵɵtemplate(8, LyCarousel_div_8_Template, 3, 4, "div", 4);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("className", ctx.classes.slide);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", ctx.lyItems.length !== 1 && ctx.hasNavigationIndicators);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.hasNavigationArrows);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.hasNavigationArrows);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.hasProgressBar && ctx._isIntervalFn && ctx.interval && ctx.autoplay);
    } }, directives: [ɵngcc3.NgIf, ɵngcc3.NgForOf, ɵngcc3.NgClass], encapsulation: 2, changeDetection: 0 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(LyCarousel, [{
        type: core.Component,
        args: [{
                selector: 'ly-carousel',
                template: "<div\n(slidestart)=\"touch && _onDragStart()\"\n(slideleft)=\"touch && _onDrag($event)\"\n(slideright)=\"touch && _onDrag($event)\"\n(slidecancel)=\"touch && _onDragCancel()\"\n(slideend)=\"touch && _onDragEnd($event)\"\n#slideContainer\n>\n  <div #_slide [className]=\"classes.slide\">\n    <ng-content></ng-content>\n  </div>\n  <ol [className]=\"classes.indicators\" *ngIf=\"lyItems.length !== 1 && hasNavigationIndicators\">\n    <li\n      *ngFor=\"let item of lyItems; index as i\"\n      [ngClass]=\"[classes.indicator, selectedIndex == i ? classes.indicatorActive : '']\"\n      (click)=\"_select(i)\"\n    >\n      <span\n        [class]=\"classes.indicatorIcon\"\n      ></span>\n    </li>\n  </ol>\n  <div *ngIf=\"hasNavigationArrows\" [ngClass]=\"[classes.actions, 'left']\" (click)=\"prev()\">\n    <svg viewBox=\"0 0 24 24\"><path d=\"M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z\"></path></svg>\n  </div>\n  <div *ngIf=\"hasNavigationArrows\" [ngClass]=\"[classes.actions, 'right']\" (click)=\"next()\">\n    <svg viewBox=\"0 0 24 24\"><path d=\"M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z\"></path></svg>\n  </div>\n  <div\n    [className]=\"classes.barContainer\"\n    *ngIf=\"hasProgressBar && _isIntervalFn && interval && autoplay\"\n  >\n    <div\n      [className]=\"classes.bar\"\n      #_progressBar\n      [style.animation-duration]=\"interval + 'ms'\"\n    ></div>\n  </div>\n</div>",
                changeDetection: core.ChangeDetectionStrategy.OnPush,
                preserveWhitespaces: false,
                host: {
                    '(mouseenter)': '_onMouseEnter()',
                    '(mouseleave)': '_onMouseLeave()'
                },
                providers: [
                    ui.StyleRenderer
                ]
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ɵngcc0.ChangeDetectorRef }, { type: ɵngcc1.LyTheme2 }, { type: ɵngcc0.Renderer2 }, { type: ɵngcc1.StyleRenderer }, { type: ɵngcc2.Platform }]; }, { mode: [{
            type: core.Input
        }], selectedIndex: [{
            type: core.Input
        }], pauseOnHover: [{
            type: core.Input
        }], touch: [{
            type: core.Input
        }], autoplay: [{
            type: core.Input
        }], hasProgressBar: [{
            type: core.Input
        }], interval: [{
            type: core.Input
        }], hasNavigationArrows: [{
            type: core.Input
        }], hasNavigationIndicators: [{
            type: core.Input
        }], slideContainer: [{
            type: core.ViewChild,
            args: ['slideContainer']
        }], _slide: [{
            type: core.ViewChild,
            args: ['_slide']
        }], _progressBar: [{
            type: core.ViewChild,
            args: ['_progressBar']
        }], lyItems: [{
            type: core.ContentChildren,
            args: [core.forwardRef(function () { return LyCarouselItem; })]
        }] }); })();
        return LyCarousel;
    }());
    LyCarousel.ctorParameters = function () { return [
        { type: core.ElementRef },
        { type: core.ChangeDetectorRef },
        { type: ui.LyTheme2 },
        { type: core.Renderer2 },
        { type: ui.StyleRenderer },
        { type: platform.Platform }
    ]; };
    LyCarousel.propDecorators = {
        slideContainer: [{ type: core.ViewChild, args: ['slideContainer',] }],
        _slide: [{ type: core.ViewChild, args: ['_slide',] }],
        _progressBar: [{ type: core.ViewChild, args: ['_progressBar',] }],
        lyItems: [{ type: core.ContentChildren, args: [core.forwardRef(function () { return LyCarouselItem; }),] }],
        mode: [{ type: core.Input }],
        selectedIndex: [{ type: core.Input }],
        pauseOnHover: [{ type: core.Input }],
        touch: [{ type: core.Input }],
        autoplay: [{ type: core.Input }],
        hasProgressBar: [{ type: core.Input }],
        interval: [{ type: core.Input }],
        hasNavigationArrows: [{ type: core.Input }],
        hasNavigationIndicators: [{ type: core.Input }]
    };
    var LyCarouselItem = /** @class */ (function () {
        function LyCarouselItem(_theme, _el) {
            this._theme = _theme;
            this._nativeElement = _el.nativeElement;
        }
        Object.defineProperty(LyCarouselItem.prototype, "srcImg", {
            set: function (value) {
                this._className = this._theme.addStyle("ly-carousel-img:" + value, ("background-image: url('" + value + "')"), this._nativeElement, this._className, STYLE_PRIORITY);
            },
            enumerable: false,
            configurable: true
        });
LyCarouselItem.ɵfac = function LyCarouselItem_Factory(t) { return new (t || LyCarouselItem)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.LyTheme2), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
LyCarouselItem.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: LyCarouselItem, selectors: [["ly-carousel-item"]], inputs: { srcImg: "srcImg" } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(LyCarouselItem, [{
        type: core.Directive,
        args: [{
                selector: 'ly-carousel-item'
            }]
    }], function () { return [{ type: ɵngcc1.LyTheme2 }, { type: ɵngcc0.ElementRef }]; }, { srcImg: [{
            type: core.Input
        }] }); })();
        return LyCarouselItem;
    }());
    LyCarouselItem.ctorParameters = function () { return [
        { type: ui.LyTheme2 },
        { type: core.ElementRef }
    ]; };
    LyCarouselItem.propDecorators = {
        srcImg: [{ type: core.Input }]
    };

    var LyCarouselModule = /** @class */ (function () {
        function LyCarouselModule() {
        }
LyCarouselModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: LyCarouselModule });
LyCarouselModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function LyCarouselModule_Factory(t) { return new (t || LyCarouselModule)(); }, imports: [[common.CommonModule, ui.LyCommonModule], ɵngcc1.LyCommonModule] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(LyCarouselModule, { declarations: [LyCarouselItem,
        LyCarousel], imports: [ɵngcc3.CommonModule, ɵngcc1.LyCommonModule], exports: [LyCarouselItem,
        LyCarousel, ɵngcc1.LyCommonModule] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(LyCarouselModule, [{
        type: core.NgModule,
        args: [{
                imports: [common.CommonModule, ui.LyCommonModule],
                exports: [LyCarouselItem, LyCarousel, ui.LyCommonModule],
                declarations: [LyCarouselItem, LyCarousel]
            }]
    }], function () { return []; }, null); })();
        return LyCarouselModule;
    }());

    /**
     * Generated bundle index. Do not edit.
     */

    exports.LyCarousel = LyCarousel;
    exports.LyCarouselItem = LyCarouselItem;
    exports.LyCarouselModule = LyCarouselModule;
    exports.STYLES = STYLES;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
