// @alyle/ui/bundles/alyle-ui-skeleton.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('@alyle/ui'),exports, require('@angular/core'), require('@alyle/ui')) :
    typeof define === 'function' && define.amd ? define('@alyle/ui/skeleton', ['@angular/core','@alyle/ui','exports', '@angular/core', '@alyle/ui'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.ng.core,global.alyle.ui,(global.ly = global.ly || {}, global.ly.skeleton = {}), global.ng.core, global.ly.core));
}(this, (function (ɵngcc0,ɵngcc1,exports, core, ui) { 'use strict';

    var STYLE_PRIORITY = -0.5;
    var STYLES = function (theme) {
        var id = ui.keyframesUniqueId.next();
        var _a = theme.background, primary = _a.primary, secondary = _a.secondary, tertiary = _a.tertiary;
        var dir = theme.direction === ui.Dir.ltr ? -1 : 1;
        var lum = primary.default.luminance();
        var one = (lum < .5
            ? tertiary
            : secondary);
        var two = (lum < .5
            ? secondary
            : tertiary);
        one = one.darken(1 * (lum < .5 ? -.5 : 0));
        two = two.darken(.25 * (lum < .5 ? -1 : 1));
        return {
            $name: LySkeleton.и,
            $priority: STYLE_PRIORITY,
            $global: function (className) { return "@keyframes " + id + "{" + className + " 0%{background-position:" + -dir * 200 + "% 50%;}" + className + " 100%{background-position:" + dir * 200 + "% 50%;}}"; },
            root: function (className) { return className + "{content:'';background:" + ("linear-gradient(270deg, " + one + ", " + two + ", " + two + ", " + one + ")") + ";background-size:400% 400%;animation:" + id + " 8s ease-in-out infinite;color:transparent;cursor:progress;user-select:none;}"; }
        };
    };
    var LySkeleton = /** @class */ (function () {
        function LySkeleton(sRenderer) {
            this.sRenderer = sRenderer;
            /** @docs-private */
            this.classes = this.sRenderer.renderSheet(STYLES);
        }
        Object.defineProperty(LySkeleton.prototype, "skeleton", {
            get: function () {
                return this._skeleton;
            },
            set: function (val) {
                var newVal = ui.toBoolean(val);
                this._skeleton = newVal;
                this.sRenderer.toggleClass(this.classes.root, newVal);
            },
            enumerable: false,
            configurable: true
        });
LySkeleton.ɵfac = function LySkeleton_Factory(t) { return new (t || LySkeleton)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.StyleRenderer)); };
LySkeleton.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: LySkeleton, selectors: [["", "lySkeleton", ""]], inputs: { skeleton: ["lySkeleton", "skeleton"] }, exportAs: ["lySkeleton"], features: [ɵngcc0.ɵɵProvidersFeature([
            ui.StyleRenderer
        ])] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(LySkeleton, [{
        type: core.Directive,
        args: [{
                selector: '[lySkeleton]',
                providers: [
                    ui.StyleRenderer
                ],
                exportAs: 'lySkeleton'
            }]
    }], function () { return [{ type: ɵngcc1.StyleRenderer }]; }, { skeleton: [{
            type: core.Input,
            args: ['lySkeleton']
        }] }); })();
        return LySkeleton;
    }());
    /** @docs-private */
    LySkeleton.и = 'LySkeleton';
    LySkeleton.ctorParameters = function () { return [
        { type: ui.StyleRenderer }
    ]; };
    LySkeleton.propDecorators = {
        skeleton: [{ type: core.Input, args: ['lySkeleton',] }]
    };

    var LySkeletonModule = /** @class */ (function () {
        function LySkeletonModule() {
        }
LySkeletonModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: LySkeletonModule });
LySkeletonModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function LySkeletonModule_Factory(t) { return new (t || LySkeletonModule)(); } });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(LySkeletonModule, { declarations: [LySkeleton], exports: [LySkeleton] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(LySkeletonModule, [{
        type: core.NgModule,
        args: [{
                declarations: [LySkeleton],
                exports: [LySkeleton]
            }]
    }], function () { return []; }, null); })();
        return LySkeletonModule;
    }());

    /**
     * Generated bundle index. Do not edit.
     */

    exports.LySkeleton = LySkeleton;
    exports.LySkeletonModule = LySkeletonModule;
    exports.STYLES = STYLES;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
