// @alyle/ui/esm2015/menu/menu.module.js
import { LyMenu, LyMenuTriggerFor, LyMenuItem, LyMenuOpenOnHover } from './menu';
import { LyCommonModule, LyOverlayModule } from '@alyle/ui';
import { FormsModule } from '@angular/forms';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
export class LyMenuModule {
}
LyMenuModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: LyMenuModule });
LyMenuModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function LyMenuModule_Factory(t) { return new (t || LyMenuModule)(); }, imports: [[CommonModule, FormsModule, LyCommonModule, LyOverlayModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(LyMenuModule, { declarations: function () { return [LyMenu, LyMenuItem, LyMenuTriggerFor, LyMenuOpenOnHover]; }, imports: function () { return [CommonModule, FormsModule, LyCommonModule, LyOverlayModule]; }, exports: function () { return [LyMenu, LyMenuItem, LyMenuTriggerFor, LyMenuOpenOnHover]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(LyMenuModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, FormsModule, LyCommonModule, LyOverlayModule],
                exports: [LyMenu, LyMenuItem, LyMenuTriggerFor, LyMenuOpenOnHover],
                declarations: [LyMenu, LyMenuItem, LyMenuTriggerFor, LyMenuOpenOnHover]
            }]
    }], null, null); })();
