// @alyle/ui/esm2015/card/card.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { LyCard, LyCardContent, LyCardActions, LyCardMedia } from './card.directive';
import { LyCommonModule } from '@alyle/ui';
import * as ɵngcc0 from '@angular/core';
export class LyCardModule {
}
LyCardModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: LyCardModule });
LyCardModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function LyCardModule_Factory(t) { return new (t || LyCardModule)(); }, imports: [[
            CommonModule
        ],
        LyCommonModule] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(LyCardModule, { declarations: function () { return [LyCard, LyCardContent, LyCardActions, LyCardMedia]; }, imports: function () { return [CommonModule]; }, exports: function () { return [LyCard, LyCardContent, LyCardActions, LyCardMedia, LyCommonModule]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(LyCardModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule
                ],
                exports: [LyCard, LyCardContent, LyCardActions, LyCardMedia, LyCommonModule],
                declarations: [LyCard, LyCardContent, LyCardActions, LyCardMedia]
            }]
    }], null, null); })();
