// @alyle/ui/esm2015/src/styles/core-styles.js
import { Injectable } from '@angular/core';
import { LyTheme2 } from '../theme/theme2.service';
import * as i0 from "@angular/core";
import * as i1 from "../theme/theme2.service";
import * as ɵngcc0 from '@angular/core';
const ɵ0 = (className) => `${className}{position:absolute;top:0;bottom:0;left:0;right:0;}`, ɵ1 = (className) => `${className}{border:0;clip:rect(0 0 0 0);height:1px;margin:-1px;overflow:hidden;padding:0;position:absolute;width:1px;outline:0;-webkit-appearance:none;-moz-appearance:none;}`, ɵ2 = (className) => `${className}{-webkit-tap-highlight-color:transparent;background-color:transparent;border:0;-moz-appearance:none;-webkit-appearance:none;margin:0;outline:none;box-sizing:border-box;position:relative;text-decoration-line:none;-webkit-text-decoration-line:none;}${className}::-moz-focus-inner:{border:0;}`;
export const LY_COMMON_STYLES = {
    fill: ɵ0,
    visuallyHidden: ɵ1,
    button: ɵ2
};
export const LY_COMMON_STYLES_DEPRECATED = {
    fill: {
        position: 'absolute',
        top: 0,
        bottom: 0,
        left: 0,
        right: 0,
    },
    visuallyHidden: {
        border: 0,
        clip: 'rect(0 0 0 0)',
        height: '1px',
        margin: '-1px',
        overflow: 'hidden',
        padding: 0,
        position: 'absolute',
        width: '1px',
        outline: 0,
        '-webkit-appearance': 'none',
        '-moz-appearance': 'none'
    },
    button: {
        '-webkit-tap-highlight-color': 'transparent',
        backgroundColor: `transparent`,
        border: 0,
        '-moz-appearance': 'none',
        '-webkit-appearance': 'none',
        margin: 0,
        outline: 'none',
        boxSizing: 'border-box',
        position: 'relative',
        textDecorationLine: 'none',
        '-webkit-text-decoration-line': 'none',
        '&::-moz-focus-inner': {
            border: 0
        }
    }
};
export class LyCoreStyles {
    constructor(theme) {
        this.theme = theme;
        this.classes = this.theme.addStyleSheet(LY_COMMON_STYLES_DEPRECATED);
    }
}
LyCoreStyles.ɵfac = function LyCoreStyles_Factory(t) { return new (t || LyCoreStyles)(ɵngcc0.ɵɵinject(LyTheme2)); };
LyCoreStyles.ɵprov = i0.ɵɵdefineInjectable({ factory: function LyCoreStyles_Factory() { return new LyCoreStyles(i0.ɵɵinject(i1.LyTheme2)); }, token: LyCoreStyles, providedIn: "root" });
LyCoreStyles.ctorParameters = () => [
    { type: LyTheme2 }
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(LyCoreStyles, [{
        type: Injectable,
        args: [{ providedIn: 'root' }]
    }], function () { return [{ type: LyTheme2 }]; }, null); })();
export { ɵ0, ɵ1, ɵ2 };
