// @alyle/ui/esm2015/carousel/carousel.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { LyCarouselItem, LyCarousel } from './carousel';
import { LyCommonModule } from '@alyle/ui';
import * as ɵngcc0 from '@angular/core';
export class LyCarouselModule {
}
LyCarouselModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: LyCarouselModule });
LyCarouselModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function LyCarouselModule_Factory(t) { return new (t || LyCarouselModule)(); }, imports: [[CommonModule, LyCommonModule],
        LyCommonModule] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(LyCarouselModule, { declarations: function () { return [LyCarouselItem, LyCarousel]; }, imports: function () { return [CommonModule, LyCommonModule]; }, exports: function () { return [LyCarouselItem, LyCarousel, LyCommonModule]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(LyCarouselModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, LyCommonModule],
                exports: [LyCarouselItem, LyCarousel, LyCommonModule],
                declarations: [LyCarouselItem, LyCarousel]
            }]
    }], null, null); })();
