// @alyle/ui/esm2015/src/dom/overlay-backdrop.js
import { Component, HostListener, ElementRef } from '@angular/core';
import { LyTheme2 } from '../theme/theme2.service';
import { LyOverlayConfig } from './overlay-config';
import { LY_COMMON_STYLES } from '../styles/core-styles';
import { StyleRenderer } from '../minimal/renderer-style';
import { st2c } from '../parse';
import * as ɵngcc0 from '@angular/core';
const STYLE_PRIORITY = -2;
export const STYLES_BACKDROP_ROOT = () => (className) => `${st2c((LY_COMMON_STYLES.fill), `${className}`)}${className}{width:100vw;height:100vh;pointer-events:all;user-select:none;}`;
export class LyOverlayBackdrop {
    constructor(_el, _theme, sRenderer, _config) {
        this._el = _el;
        this.sRenderer = sRenderer;
        this._config = _config;
        _el.nativeElement.classList.add(sRenderer.add(STYLES_BACKDROP_ROOT, STYLE_PRIORITY));
        // this applies custom class for backdrop,
        // if one is not defined, do nothing.
        const backdropClass = _config.backdropClass;
        if (backdropClass) {
            this._el.nativeElement.classList.add(backdropClass);
        }
    }
    onclick() {
        if (!this._config.disableClose) {
            this._config.fnDestroy();
        }
    }
}
LyOverlayBackdrop.ɵfac = function LyOverlayBackdrop_Factory(t) { return new (t || LyOverlayBackdrop)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(LyTheme2), ɵngcc0.ɵɵdirectiveInject(StyleRenderer), ɵngcc0.ɵɵdirectiveInject(LyOverlayConfig)); };
LyOverlayBackdrop.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: LyOverlayBackdrop, selectors: [["ly-overlay-backdrop"]], hostBindings: function LyOverlayBackdrop_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("click", function LyOverlayBackdrop_click_HostBindingHandler() { return ctx.onclick(); });
    } }, features: [ɵngcc0.ɵɵProvidersFeature([
            StyleRenderer
        ])], decls: 0, vars: 0, template: function LyOverlayBackdrop_Template(rf, ctx) { }, encapsulation: 2 });
LyOverlayBackdrop.ctorParameters = () => [
    { type: ElementRef },
    { type: LyTheme2 },
    { type: StyleRenderer },
    { type: LyOverlayConfig }
];
LyOverlayBackdrop.propDecorators = {
    onclick: [{ type: HostListener, args: ['click',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(LyOverlayBackdrop, [{
        type: Component,
        args: [{
                selector: 'ly-overlay-backdrop',
                template: ``,
                providers: [
                    StyleRenderer
                ]
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: LyTheme2 }, { type: StyleRenderer }, { type: LyOverlayConfig }]; }, { onclick: [{
            type: HostListener,
            args: ['click']
        }] }); })();
