// @alyle/ui/esm2015/src/dom/overlay-container.js
import { Injectable } from '@angular/core';
import { LyTheme2 } from '../theme/theme2.service';
import { Platform } from '@angular/cdk/platform';
import * as i0 from "@angular/core";
import * as i1 from "../theme/theme2.service";
import * as i2 from "@angular/cdk/platform";
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/cdk/platform';
const styles = (theme) => ({
    overlay: {
        position: 'fixed',
        top: 0,
        left: 0,
        right: 0,
        bottom: 0,
        zIndex: theme.zIndex.overlay,
        pointerEvents: 'none'
    }
});
const ɵ0 = styles;
export class LyOverlayContainer {
    constructor(theme, _platform) {
        this.theme = theme;
        this._platform = _platform;
        this._classes = this.theme.addStyleSheet(styles);
        this._items = new Set();
        if (this._platform.isBrowser) {
            const container = document.createElement('ly-overlay-container');
            document.body.appendChild(container);
            this._containerElement = container;
        }
    }
    get overlayLen() {
        return this._items.size;
    }
    get containerElement() {
        return this._containerElement;
    }
    /**
     * Add instance
     * @ignore
     */
    _add(item, elementAfter) {
        this._items.add(item);
        if ((elementAfter === null || elementAfter === void 0 ? void 0 : elementAfter.parentElement) && item.parentElement === elementAfter.parentElement) {
            this.containerElement.insertBefore(item, elementAfter);
        }
        else {
            this.containerElement.appendChild(item);
        }
        this._update();
    }
    /**
   * Remove instance
   * @ignore
   */
    _remove(item) {
        this.containerElement.removeChild(item);
        this._items.delete(item);
        this._update();
    }
    /**
     * Update styles for overlay container
     * @ignore
     */
    _update() {
        if (this._items.size) {
            if (!this._isActiveOverlayContainer) {
                this._isActiveOverlayContainer = true;
                this._containerElement.classList.add(this._classes.overlay);
            }
        }
        else if (this._isActiveOverlayContainer) {
            this._containerElement.classList.remove(this._classes.overlay);
            this._isActiveOverlayContainer = false;
        }
    }
}
LyOverlayContainer.ɵfac = function LyOverlayContainer_Factory(t) { return new (t || LyOverlayContainer)(ɵngcc0.ɵɵinject(LyTheme2), ɵngcc0.ɵɵinject(ɵngcc1.Platform)); };
LyOverlayContainer.ɵprov = i0.ɵɵdefineInjectable({ factory: function LyOverlayContainer_Factory() { return new LyOverlayContainer(i0.ɵɵinject(i1.LyTheme2), i0.ɵɵinject(i2.Platform)); }, token: LyOverlayContainer, providedIn: "root" });
LyOverlayContainer.ctorParameters = () => [
    { type: LyTheme2 },
    { type: Platform }
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(LyOverlayContainer, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: LyTheme2 }, { type: ɵngcc1.Platform }]; }, null); })();
export { ɵ0 };
