// @alyle/ui/esm2015/tabs/tabs.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { NgTranscludeModule, LyCommonModule, LyThemeModule } from '@alyle/ui';
import { LyTabs, LyTabLabel, LyTab } from './tabs';
import { LyTabContent } from './tab-content.directive';
import { PortalModule } from '@angular/cdk/portal';
import * as ɵngcc0 from '@angular/core';
export class LyTabsModule {
}
LyTabsModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: LyTabsModule });
LyTabsModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function LyTabsModule_Factory(t) { return new (t || LyTabsModule)(); }, imports: [[
            LyThemeModule,
            CommonModule,
            LyCommonModule,
            NgTranscludeModule,
            PortalModule
        ],
        LyCommonModule] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(LyTabsModule, { declarations: function () { return [LyTabs, LyTab, LyTabLabel, LyTabContent]; }, imports: function () { return [LyThemeModule,
        CommonModule,
        LyCommonModule,
        NgTranscludeModule,
        PortalModule]; }, exports: function () { return [LyCommonModule, LyTabs, LyTab, LyTabLabel, LyTabContent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(LyTabsModule, [{
        type: NgModule,
        args: [{
                imports: [
                    LyThemeModule,
                    CommonModule,
                    LyCommonModule,
                    NgTranscludeModule,
                    PortalModule
                ],
                exports: [LyCommonModule, LyTabs, LyTab, LyTabLabel, LyTabContent],
                declarations: [LyTabs, LyTab, LyTabLabel, LyTabContent]
            }]
    }], null, null); })();
