// @alyle/ui/esm2015/divider/divider.js
import { Directive, Input } from '@angular/core';
import { StyleRenderer, toBoolean } from '@alyle/ui';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@alyle/ui';
const STYLES = (theme) => (className) => `${className}{display:block;background-color:${theme.divider};height:1px;}`;
const ɵ0 = STYLES;
export class LyDivider {
    constructor(sRenderer) {
        this.sRenderer = sRenderer;
    }
    /** Add indentation (72px) */
    set inset(val) {
        const newVal = this._inset = toBoolean(val);
        if (newVal) {
            this[0x1] = this.sRenderer.add(`${LyDivider.и}--inset`, ({ before }) => (className) => `${className}{margin-${before}:74px;}`, this[0x1]);
        }
        else {
            this.sRenderer.removeClass(this[0x1]);
        }
    }
    get inset() {
        return this._inset;
    }
    ngOnInit() {
        this.sRenderer.add(STYLES);
    }
}
LyDivider.ɵfac = function LyDivider_Factory(t) { return new (t || LyDivider)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.StyleRenderer)); };
LyDivider.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: LyDivider, selectors: [["ly-divider"]], inputs: { inset: "inset" }, features: [ɵngcc0.ɵɵProvidersFeature([
            StyleRenderer
        ])] });
LyDivider.и = 'LyDivider';
LyDivider.ctorParameters = () => [
    { type: StyleRenderer }
];
LyDivider.propDecorators = {
    inset: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(LyDivider, [{
        type: Directive,
        args: [{
                selector: 'ly-divider',
                providers: [
                    StyleRenderer
                ]
            }]
    }], function () { return [{ type: ɵngcc1.StyleRenderer }]; }, { inset: [{
            type: Input
        }] }); })();
export { ɵ0 };
