// @alyle/ui/esm2015/dialog/dialog.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { LyCommonModule, LyOverlayModule } from '@alyle/ui';
import { LyDialogContainer } from './dialog-container.component';
import { LyDialog } from './dialog';
import { LyDialogTitle } from './dialog-title.directive';
import { LyDialogContent } from './dialog-content.directive';
import { LyDialogActions } from './dialog-actions.directive';
import * as ɵngcc0 from '@angular/core';
export class LyDialogModule {
}
LyDialogModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: LyDialogModule });
LyDialogModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function LyDialogModule_Factory(t) { return new (t || LyDialogModule)(); }, providers: [
        LyDialog
    ], imports: [[
            CommonModule,
            LyCommonModule,
            LyOverlayModule
        ],
        LyCommonModule] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(LyDialogModule, { declarations: function () { return [LyDialogContainer,
        LyDialogTitle,
        LyDialogContent,
        LyDialogActions]; }, imports: function () { return [CommonModule,
        LyCommonModule,
        LyOverlayModule]; }, exports: function () { return [LyCommonModule,
        LyDialogContainer,
        LyDialogTitle,
        LyDialogContent,
        LyDialogActions]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(LyDialogModule, [{
        type: NgModule,
        args: [{
                entryComponents: [
                    LyDialogContainer
                ],
                declarations: [
                    LyDialogContainer,
                    LyDialogTitle,
                    LyDialogContent,
                    LyDialogActions
                ],
                imports: [
                    CommonModule,
                    LyCommonModule,
                    LyOverlayModule
                ],
                exports: [
                    LyCommonModule,
                    LyDialogContainer,
                    LyDialogTitle,
                    LyDialogContent,
                    LyDialogActions
                ],
                providers: [
                    LyDialog
                ]
            }]
    }], null, null); })();
