// @alyle/ui/esm2015/dialog/dialog-ref.js
import { Injectable } from '@angular/core';
import { LyOverlayRef } from '@alyle/ui';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@alyle/ui';
export class LyDialogRef {
    constructor(_overlayRef) {
        this._overlayRef = _overlayRef;
    }
    get afterOpened() {
        return this._overlayRef.componentRef.instance._afterOpened.asObservable();
    }
    get beforeClosed() {
        return this._overlayRef.componentRef.instance._beforeClosed.asObservable();
    }
    get afterClosed() {
        return this._overlayRef.componentRef.instance._afterClosed.asObservable();
    }
    /**
     * @internal
     * @docs-private
     */
    get result() {
        return this._result;
    }
    close(result) {
        const dialogContainer = this._overlayRef.componentRef.instance;
        dialogContainer._beforeClosed.next(result);
        dialogContainer._beforeClosed.complete();
        dialogContainer._startClose();
        this._result = result;
    }
}
LyDialogRef.ɵfac = function LyDialogRef_Factory(t) { return new (t || LyDialogRef)(ɵngcc0.ɵɵinject(ɵngcc1.LyOverlayRef)); };
LyDialogRef.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: LyDialogRef, factory: LyDialogRef.ɵfac });
LyDialogRef.ctorParameters = () => [
    { type: LyOverlayRef }
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(LyDialogRef, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.LyOverlayRef }]; }, null); })();
