// @alyle/ui/esm2015/slider/mark.js
import { untilComponentDestroyed } from '@alyle/ui';
import { Component, Input, Renderer2, ElementRef, ChangeDetectionStrategy, ViewChild } from '@angular/core';
import { LySlider, гvalueToPercent, гbetween } from './slider';
import { LyTick } from './tick';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './tick';

const _c0 = ["*"];
export class LyMark {
    constructor(_slider, _renderer, _el) {
        this._slider = _slider;
        this._renderer = _renderer;
        this._el = _el;
        /** @docs-private */
        this.classes = this._slider.classes;
        _renderer.addClass(_el.nativeElement, _slider.classes.mark);
    }
    ngOnInit() {
        this._renderer.insertBefore(this._slider._getHostElement(), this._tick._getHostElement(), this._slider._ticksRef.nativeElement);
        this._slider._changes.pipe(untilComponentDestroyed(this)).subscribe(() => {
            this._updateMark();
        });
    }
    _updateMark() {
        const min = this._slider._minPercent;
        const max = this._slider._maxPercent;
        const className = this._slider.classes.markActive;
        const percent = гvalueToPercent(this.value, this._slider.min, this._slider.max);
        const pos = this._slider._calculatePosition(percent);
        if (гbetween(percent, min, max)) {
            this._markActiveClass = className;
            this._renderer.addClass(this._el.nativeElement, className);
        }
        else if (this._markActiveClass) {
            this._markActiveClass = null;
            this._renderer.removeClass(this._el.nativeElement, className);
        }
        this._renderer.setStyle(this._getHostElement(), 'bottom', null);
        this._renderer.setStyle(this._getHostElement(), 'left', null);
        this._renderer.setStyle(this._getHostElement(), 'right', null);
        this._renderer.setStyle(this._getHostElement(), pos.style, pos.value);
    }
    ngOnDestroy() { }
    _getHostElement() {
        return this._el.nativeElement;
    }
}
LyMark.ɵfac = function LyMark_Factory(t) { return new (t || LyMark)(ɵngcc0.ɵɵdirectiveInject(LySlider), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Renderer2), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
LyMark.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: LyMark, selectors: [["ly-mark"]], viewQuery: function LyMark_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵstaticViewQuery(LyTick, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx._tick = _t.first);
    } }, inputs: { value: "value" }, ngContentSelectors: _c0, decls: 2, vars: 1, consts: [[3, "value"]], template: function LyMark_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵelement(0, "ly-tick", 0);
        ɵngcc0.ɵɵprojection(1);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("value", ctx.value);
    } }, directives: [ɵngcc1.LyTick], encapsulation: 2, changeDetection: 0 });
LyMark.ctorParameters = () => [
    { type: LySlider },
    { type: Renderer2 },
    { type: ElementRef }
];
LyMark.propDecorators = {
    _tick: [{ type: ViewChild, args: [LyTick, { static: true },] }],
    value: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(LyMark, [{
        type: Component,
        args: [{
                selector: 'ly-mark',
                template: "<ly-tick [value]=\"value\"></ly-tick>\n<ng-content></ng-content>",
                changeDetection: ChangeDetectionStrategy.OnPush
            }]
    }], function () { return [{ type: LySlider }, { type: ɵngcc0.Renderer2 }, { type: ɵngcc0.ElementRef }]; }, { _tick: [{
            type: ViewChild,
            args: [LyTick, { static: true }]
        }], value: [{
            type: Input
        }] }); })();
