// @alyle/ui/esm2015/slider/slider.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { LyCommonModule } from '@alyle/ui';
import { LySlider } from './slider';
import { LyMark } from './mark';
import { LyTick } from './tick';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from '@alyle/ui';
export class LySliderModule {
}
LySliderModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: LySliderModule });
LySliderModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function LySliderModule_Factory(t) { return new (t || LySliderModule)(); }, imports: [[
            CommonModule,
            LyCommonModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(LySliderModule, { declarations: function () { return [LySlider, LyMark, LyTick]; }, imports: function () { return [CommonModule,
        LyCommonModule]; }, exports: function () { return [LySlider, LyMark]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(LySliderModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    LyCommonModule
                ],
                declarations: [LySlider, LyMark, LyTick],
                exports: [LySlider, LyMark]
            }]
    }], null, null); })();
ɵngcc0.ɵɵsetComponentScope(LySlider, [ɵngcc1.NgClass, ɵngcc1.NgComponentOutlet, ɵngcc1.NgForOf, ɵngcc1.NgIf, ɵngcc1.NgTemplateOutlet, ɵngcc1.NgStyle, ɵngcc1.NgSwitch, ɵngcc1.NgSwitchCase, ɵngcc1.NgSwitchDefault, ɵngcc1.NgPlural, ɵngcc1.NgPluralCase, ɵngcc2.LyStyle, ɵngcc2.ɵa, ɵngcc2.LyPaper, LySlider, LyMark, LyTick], [ɵngcc1.AsyncPipe, ɵngcc1.UpperCasePipe, ɵngcc1.LowerCasePipe, ɵngcc1.JsonPipe, ɵngcc1.SlicePipe, ɵngcc1.DecimalPipe, ɵngcc1.PercentPipe, ɵngcc1.TitleCasePipe, ɵngcc1.CurrencyPipe, ɵngcc1.DatePipe, ɵngcc1.I18nPluralPipe, ɵngcc1.I18nSelectPipe, ɵngcc1.KeyValuePipe]);
